From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Tue, 19 Mar 2019 22:11:14 +0200
Subject: [PATCH] Improve configuration adapter getting/setting API


diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyServer.java b/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
index 218ec45c922b1926651b42571db111685abe8911..40527301750000d942ff33d81e82f51b82a50870 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
@@ -136,6 +136,7 @@ public abstract class ProxyServer
      *
      * @return the used configuration adapter
      */
+    @Deprecated // MikroCord
     public abstract ConfigurationAdapter getConfigurationAdapter();
 
     /**
@@ -144,8 +145,28 @@ public abstract class ProxyServer
      *
      * @param adapter the adapter to use
      */
+    @Deprecated // MikroCord
     public abstract void setConfigurationAdapter(ConfigurationAdapter adapter);
 
+    // MikroCord start - Configuration adapters
+    /**
+     * Returns the currently in use configuration adapter by {@link ConfigurationType}.
+     *
+     * @param configurationType The configuration type
+     * @return the used configuration adapter
+     */
+    public abstract ConfigurationAdapter getConfigurationAdapter(ConfigurationType configurationType);
+
+    /**
+     * Set the configuration adapter to be used with given {@link ConfigurationType}. Must be called from
+     * {@link Plugin#onLoad()}.
+     *
+     * @param adapter the adapter to use
+     * @param configurationType The configuration type
+     */
+    public abstract void setConfigurationAdapter(ConfigurationType configurationType, ConfigurationAdapter adapter);
+    // MikroCord end
+
     /**
      * Get the currently in use reconnect handler.
      *
@@ -314,4 +335,18 @@ public abstract class ProxyServer
      */
     public abstract Title createTitle();
 
+    // MikroCord start - Configuration adapters
+    public enum ConfigurationType {
+        /**
+         * Generic BungeeCord configuration
+         */
+        BUNGEECORD,
+
+        /**
+         * Waterfall-specific configuration
+         */
+        WATERFALL,
+        ;
+    }
+    // MikroCord end
 }
diff --git a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
index e860214f8d7bedcc98d5f4cc0c2108ea153b80be..b6840f65f336d0c5ed3f8b610fde2b17bbc2512a 100644
--- a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
+++ b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
@@ -47,8 +47,13 @@ public class WaterfallConfiguration extends Configuration {
     @Override
     public void load() {
         super.load();
+        /* // MikroCord start - Improve configuration adapter getting/setting API
         YamlConfig config = new YamlConfig(new File("waterfall.yml"));
         config.load(false); // Load, but no permissions
+        */
+        var config = net.md_5.bungee.api.ProxyServer.getInstance().getConfigurationAdapter(net.md_5.bungee.api.ProxyServer.ConfigurationType.WATERFALL);
+        config.load();
+        // MikroCord end
         logInitialHandlerConnections = config.getBoolean( "log_initial_handler_connections", logInitialHandlerConnections );
         gameVersion = config.getString("game_version", "").isEmpty() ? Joiner.on(", ").join(ProtocolConstants.SUPPORTED_VERSIONS) : config.getString("game_version", "");
         useNettyDnsResolver = config.getBoolean("use_netty_dns_resolver", useNettyDnsResolver);
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index b69720ce36f2dc01f42c8683832b51c122b84e24..ee3bb4ce9c65992a62878d77285928b24426cd42 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -166,6 +166,37 @@ public class BungeeCord extends ProxyServer
     private ConnectionThrottle connectionThrottle;
     //private final ModuleManager moduleManager = new ModuleManager(); // MikroCord - drop modules support
 
+    // MikroCord start - Configuration adapters
+    private ConfigurationAdapter waterfallConfigurationAdapter = new YamlConfig(new File("waterfall.yml")) {
+        @Override
+        public void load() {
+            super.load(false);
+        }
+    };
+
+    @Override
+    public ConfigurationAdapter getConfigurationAdapter(ConfigurationType configurationType) {
+        switch (configurationType) {
+            case BUNGEECORD:
+                return configurationAdapter;
+            case WATERFALL:
+                return waterfallConfigurationAdapter;
+        }
+        return null;
+    }
+
+    @Override
+    public void setConfigurationAdapter(ConfigurationType configurationType, ConfigurationAdapter adapter) {
+        switch (configurationType) {
+            case BUNGEECORD:
+                configurationAdapter = adapter;
+                break;
+            case WATERFALL:
+                waterfallConfigurationAdapter = adapter;
+                break;
+        }
+    }
+    // MikroCord end
     {
         // TODO: Proper fallback when we interface the manager
         registerChannel( "BungeeCord" );
