From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Tue, 19 Mar 2019 23:05:17 +0200
Subject: [PATCH] Add support for MikroCord-specific configuration


diff --git a/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java b/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java
new file mode 100644
index 0000000000000000000000000000000000000000..be3bbe9adf29253ec1d62ac7e563629e0ea7f9a5
--- /dev/null
+++ b/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java
@@ -0,0 +1,7 @@
+package eu.mikroskeem.mikrocord.api.config;
+
+/**
+ * @author Mark Vainomaa
+ */
+public interface MikroCordProxyConfig {
+}
diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
index 9f8f474079b74ed3bb590799e16bd8ddec7cfe54..647d2493f6fdfe7297124715d2f0165fb21a528e 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
@@ -11,7 +11,7 @@ import net.md_5.bungee.api.config.ServerInfo;
  * @deprecated This class is subject to rapid change between releases
  */
 @Deprecated
-public interface ProxyConfig
+public interface ProxyConfig extends eu.mikroskeem.mikrocord.api.config.MikroCordProxyConfig
 {
 
     /**
diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyServer.java b/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
index 40527301750000d942ff33d81e82f51b82a50870..d40489cb1c2346b1dfbdd3e17ef1e23950be98ba 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
@@ -346,6 +346,11 @@ public abstract class ProxyServer
          * Waterfall-specific configuration
          */
         WATERFALL,
+
+        /**
+         * MikroCord-specific configuration
+         */
+        MIKROCORD,
         ;
     }
     // MikroCord end
diff --git a/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
new file mode 100644
index 0000000000000000000000000000000000000000..0b613f678092224ee72152a718a96fe7702ab6ef
--- /dev/null
+++ b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
@@ -0,0 +1,17 @@
+package eu.mikroskeem.mikrocord.conf;
+
+import io.github.waterfallmc.waterfall.conf.WaterfallConfiguration;
+import net.md_5.bungee.api.ProxyServer;
+import net.md_5.bungee.api.config.ConfigurationAdapter;
+
+/**
+ * @author Mark Vainomaa
+ */
+public class MikroCordConfiguration extends WaterfallConfiguration {
+    @Override
+    public void load() {
+        super.load();
+        ConfigurationAdapter config = ProxyServer.getInstance().getConfigurationAdapter(ProxyServer.ConfigurationType.MIKROCORD);
+        config.load();
+    }
+}
diff --git a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
index b6840f65f336d0c5ed3f8b610fde2b17bbc2512a..b1adfdd0c547befd93fef12322b29c26d23e2779 100644
--- a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
+++ b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
@@ -7,7 +7,7 @@ import net.md_5.bungee.protocol.ProtocolConstants;
 
 import java.io.File;
 
-public class WaterfallConfiguration extends Configuration {
+public abstract class WaterfallConfiguration extends Configuration { // MikroCord - make abstract
 
     /**
      * Whether we log InitialHandler connections
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index ee3bb4ce9c65992a62878d77285928b24426cd42..334a608b4947d0fd426b4574c6c94a1007e85935 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -105,7 +105,7 @@ public class BungeeCord extends ProxyServer
      * Configuration.
      */
     @Getter
-    public final Configuration config = new WaterfallConfiguration();
+    public final Configuration config = new eu.mikroskeem.mikrocord.conf.MikroCordConfiguration(); // MikroCord - Fork specific configuration
     /**
      * Localization bundle.
      */
@@ -173,6 +173,14 @@ public class BungeeCord extends ProxyServer
             super.load(false);
         }
     };
+    // MikroCord start - Fork specific configuration
+    private ConfigurationAdapter mikrocordConfigurationAdapter = new YamlConfig(new File("mikrocord.yml")) {
+        @Override
+        public void load() {
+            super.load(false);
+        }
+    };
+    // MikroCord end
 
     @Override
     public ConfigurationAdapter getConfigurationAdapter(ConfigurationType configurationType) {
@@ -181,6 +189,10 @@ public class BungeeCord extends ProxyServer
                 return configurationAdapter;
             case WATERFALL:
                 return waterfallConfigurationAdapter;
+            // Mikrocord start - Fork specific configuration
+            case MIKROCORD:
+                return mikrocordConfigurationAdapter;
+            // MikroCord end
         }
         return null;
     }
@@ -194,6 +206,11 @@ public class BungeeCord extends ProxyServer
             case WATERFALL:
                 waterfallConfigurationAdapter = adapter;
                 break;
+            // Mikrocord start - Fork specific configuration
+            case MIKROCORD:
+                mikrocordConfigurationAdapter = adapter;
+                break;
+            // MikroCord end
         }
     }
     // MikroCord end
