From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Wed, 20 Mar 2019 07:48:35 +0200
Subject: [PATCH] Get rid of startup delay


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
index 49f1ac8e8d3dffb725793709ddbf48b2444df4af..adbe8f37ecd8421843bc490916a14e6b848c4c4f 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
@@ -48,11 +48,15 @@ public class BungeeCordLauncher
             deadline.add( Calendar.WEEK_OF_YEAR, -8 );
             if ( buildDate.before( deadline.getTime() ) )
             {
+                /* Firefly start - remove this dumb delay
                 System.err.println( "*** Hey! This build is potentially outdated :( ***" );
                 System.err.println( "*** Please check for a new build from https://papermc.io/ci/job/Waterfall/ ***" );
                 System.err.println( "*** Should this build be outdated, you will get NO support for it. ***" );
                 System.err.println( "*** Server will start in 10 seconds ***" );
                 Thread.sleep( TimeUnit.SECONDS.toMillis( 10 ) );
+                */
+                System.err.println("This Firefly build is potentially outdated, please check for updates");
+                // Firefly end
             }
         }
 
