From 3058f5fbf141fc1fee51751a532a4b46754f925f Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Wed, 20 Mar 2019 07:48:35 +0200
Subject: [PATCH] Get rid of startup delay


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
index 624eb5f0..6caa06fc 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
@@ -48,11 +48,15 @@ public class BungeeCordLauncher
             deadline.add( Calendar.WEEK_OF_YEAR, -4 );
             if ( buildDate.before( deadline.getTime() ) )
             {
+                /* MikroCord start - remove this dumb delay
                 System.err.println( "*** Hey! This build is potentially outdated :( ***" );
                 System.err.println( "*** Please check for a new build from https://papermc.io/ci/job/Waterfall/ ***" );
                 System.err.println( "*** Should this build be outdated, you will get NO support for it. ***" );
                 System.err.println( "*** Server will start in 10 seconds ***" );
                 Thread.sleep( TimeUnit.SECONDS.toMillis( 10 ) );
+                */
+                System.err.println("This MikroCord build is potentially outdated, please check for updates");
+                // MikroCord end
             }
         }
 
-- 
2.21.0

