From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sat, 13 Apr 2019 21:01:27 +0300
Subject: [PATCH] Get rid of dumb Guava Subscribe annotation warning


diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
index 51559a9682e295313d7e9f3f14c6c3c7433020dc..9105530acb8c091986e304f9673b2b88e31e608b 100644
--- a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
@@ -456,11 +456,13 @@ public class PluginManager
      */
     public void registerListener(Plugin plugin, Listener listener)
     {
+        /* // MikroCord start - get rid of dumb Guava Subscribe annotation warning
         for ( Method method : listener.getClass().getDeclaredMethods() )
         {
             Preconditions.checkArgument( !method.isAnnotationPresent( Subscribe.class ),
                     "Listener %s has registered using deprecated subscribe annotation! Please update to @EventHandler.", listener );
         }
+        */ // MikroCord end
         eventBus.register( listener );
         listenersByPlugin.put( plugin, listener );
     }
