From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sat, 13 Apr 2019 23:10:17 +0300
Subject: [PATCH] Set sane default configuration values


diff --git a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
index b1adfdd0c547befd93fef12322b29c26d23e2779..93c7f7ef2f6f3a470995692787d5210332ebdec9 100644
--- a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
+++ b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
@@ -14,7 +14,7 @@ public abstract class WaterfallConfiguration extends Configuration { // MikroCor
      * <p>
      * Default is true
      */
-    private boolean logInitialHandlerConnections = true;
+    private boolean logInitialHandlerConnections = false; // MikroCord - sane configuration defaults
 
     /**
      * The supported versions displayed to the client
diff --git a/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java b/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
index 0ca9d16a6b487b362f7d5f4228c859afa377ff75..fd60b33e694e89ebf9957ee16d5d62d663989eb8 100644
--- a/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
+++ b/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
@@ -39,7 +39,7 @@ public abstract class Configuration implements ProxyConfig
     /**
      * UUID used for metrics.
      */
-    private String uuid = UUID.randomUUID().toString();
+    private String uuid; /*= UUID.randomUUID().toString();*/ // MikroCord - sane configuration defaults
     /**
      * Set of all listeners.
      */
@@ -57,7 +57,7 @@ public abstract class Configuration implements ProxyConfig
      * Whether we log proxy commands to the proxy log
      */
     private boolean logCommands;
-    private boolean logPings = true;
+    private boolean logPings = false; // MikroCord - sane configuration defaults
     private int remotePingCache = -1;
     private int playerLimit = -1;
     private Collection<String> disabledCommands;
@@ -65,7 +65,7 @@ public abstract class Configuration implements ProxyConfig
     private int remotePingTimeout = 5000;
     private int throttle = 4000;
     private int throttleLimit = 3;
-    private boolean ipForward;
+    private boolean ipForward = true; // MikroCord - sane configuration defaults
     private Favicon favicon;
     private int compressionThreshold = 256;
     private boolean preventProxyConnections;
@@ -91,7 +91,7 @@ public abstract class Configuration implements ProxyConfig
 
         listeners = adapter.getListeners();
         timeout = adapter.getInt( "timeout", timeout );
-        uuid = adapter.getString( "stats", uuid );
+        uuid = adapter.getString( "stats", () -> UUID.randomUUID().toString() ); // MikroCord - sane configuration defaults
         onlineMode = adapter.getBoolean( "online_mode", onlineMode );
         logCommands = adapter.getBoolean( "log_commands", logCommands );
         logPings = adapter.getBoolean( "log_pings", logPings );
diff --git a/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java b/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
index 549e59ef1dff844df0f8f53fbd5b94d69ce7aaf1..908ceee5a92233c5790400ffd8a835bd4a5610cf 100644
--- a/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
+++ b/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
@@ -110,7 +110,11 @@ public class YamlConfig implements ConfigurationAdapter
         Map<String, Object> groups = get( "groups", null );
         if ( groups == null )
         {
+            /* // MikroCord start - sane configuration defaults
             set( "groups.md_5", Collections.singletonList( "admin" ) );
+            */
+            set("groups", Collections.emptyMap()); // MikroCord - sane configuration defaults
+            // MikroCord end
         }
     }
 
@@ -244,7 +248,7 @@ public class YamlConfig implements ConfigurationAdapter
             new HashMap()
         } ) );
         Map<String, String> forcedDef = new HashMap<>();
-        forcedDef.put( "pvp.md-5.net", "pvp" );
+        //forcedDef.put( "pvp.md-5.net", "pvp" ); MikroCord - sane configuration defaults
 
         Collection<ListenerInfo> ret = new HashSet<>();
 
