From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sun, 14 Apr 2019 17:18:40 +0300
Subject: [PATCH] Fix potential race condition


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 9b617bf324e058562904d907161dc90b1589308f..f42e4583b95567067da075a92b8614e9ef1d627f 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -648,7 +648,14 @@ public class BungeeCord extends ProxyServer
     @Override
     public int getOnlineCount()
     {
+        // Firefly start - fix potential race condition
+        connectionLock.readLock().lock();
+        try {
         return connections.size();
+        } finally {
+            connectionLock.readLock().unlock();
+        }
+        // Firefly end
     }
 
     @Override
