From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sun, 14 Apr 2019 17:18:40 +0300
Subject: [PATCH] Fix potential race condition


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index d8b574e797a9a7356d3b3080779cd8ec8149a7e2..006b5e834b97f1b99dfdc6075ec246e15c68c37e 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -608,7 +608,14 @@ public class BungeeCord extends ProxyServer
     @Override
     public int getOnlineCount()
     {
+        // Firefly start - fix potential race condition
+        connectionLock.readLock().lock();
+        try {
         return connections.size();
+        } finally {
+            connectionLock.readLock().unlock();
+        }
+        // Firefly end
     }
 
     @Override
