From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sun, 14 Apr 2019 17:18:40 +0300
Subject: [PATCH] Fix potential race condition


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 701c3461b309e2312ca2f8b7e8fe2ea8ac4e8924..b8ac5b909faf1d1829f1145c278d6b418b4f157b 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -608,7 +608,14 @@ public class BungeeCord extends ProxyServer
     @Override
     public int getOnlineCount()
     {
+        // MikroCord start - fix potential race condition
+        connectionLock.readLock().lock();
+        try {
         return connections.size();
+        } finally {
+            connectionLock.readLock().unlock();
+        }
+        // MikroCord end
     }
 
     @Override
