From d6142dba7fd27d1dd7fe169d513247bef175888c Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sun, 14 Apr 2019 18:38:49 +0300
Subject: [PATCH] Implement part of SpigotMC/BungeeCord#1744


diff --git a/api/src/main/java/eu/mikroskeem/mikrocord/api/command/CommandExecutionException.java b/api/src/main/java/eu/mikroskeem/mikrocord/api/command/CommandExecutionException.java
new file mode 100644
index 00000000..33314574
--- /dev/null
+++ b/api/src/main/java/eu/mikroskeem/mikrocord/api/command/CommandExecutionException.java
@@ -0,0 +1,22 @@
+package eu.mikroskeem.mikrocord.api.command;
+
+/**
+ * @author Mark Vainomaa
+ */
+public class CommandExecutionException extends RuntimeException {
+    public CommandExecutionException() {
+
+    }
+
+    public CommandExecutionException(String message) {
+        super(message);
+    }
+
+    public CommandExecutionException(String message, Throwable cause) {
+        super(message, cause);
+    }
+
+    public CommandExecutionException(Throwable cause) {
+        super(cause);
+    }
+}
diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
index 88c4328f..4e1989e9 100644
--- a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
@@ -152,6 +152,23 @@ public class PluginManager
      * @return whether the command was handled
      */
     public boolean dispatchCommand(CommandSender sender, String commandLine, List<String> tabResults)
+    // MikroCord start - Import SpigotMC/BungeeCord#1744
+    {
+        return dispatchCommand(sender, commandLine, tabResults, false);
+    }
+
+    /**
+     * Execute a command if it is registered, else return false.
+     *
+     * @param sender the sender executing the command
+     * @param commandLine the complete command line including command name and
+     * arguments
+     * @param tabResults the List for tab results to be added to, or null if executing the command
+     * @param throwOnFailure whether or not to throw an {@link eu.mikroskeem.mikrocord.api.command.CommandExecutionException} if the command executed throws an exception
+     * @return whether the command was handled
+     */
+    public boolean dispatchCommand(CommandSender sender, String commandLine, List<String> tabResults, boolean throwOnFailure)
+    // MikroCord end
     {
         String[] split = commandLine.split( " ", -1 );
         // Check for chat that only contains " "
@@ -199,6 +216,7 @@ public class PluginManager
         {
             sender.sendMessage( ChatColor.RED + "An internal error occurred whilst executing this command, please check the console log for details." );
             ProxyServer.getInstance().getLogger().log( Level.WARNING, "Error in dispatching command", ex );
+            if(throwOnFailure) throw new eu.mikroskeem.mikrocord.api.command.CommandExecutionException(ex); // MikroCord - Import SpigotMC/BungeeCord#1744
         }
         return true;
     }
-- 
2.21.0

