From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sun, 14 Apr 2019 20:47:31 +0300
Subject: [PATCH] Wrap localization into configuration API


diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyServer.java b/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
index 525a8071cbb9c5c6d3f73f587c7bdb1ae678654e..54a1e5094ea5212e86be26832389d427824153d9 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyServer.java
@@ -355,6 +355,11 @@ public abstract class ProxyServer
          */
         BUNGEECORD,
 
+        /**
+         * BungeeCord messages
+         */
+        BUNGEECORD_MESSAGES,
+
         /**
          * Waterfall-specific configuration
          */
diff --git a/proxy/src/main/java/eu/mikroskeem/firefly/misc/ResourceBundleConfigurationWrapper.java b/proxy/src/main/java/eu/mikroskeem/firefly/misc/ResourceBundleConfigurationWrapper.java
new file mode 100644
index 0000000000000000000000000000000000000000..6ee1f37b91f9962b0e8e390e17b3fd1f02662538
--- /dev/null
+++ b/proxy/src/main/java/eu/mikroskeem/firefly/misc/ResourceBundleConfigurationWrapper.java
@@ -0,0 +1,61 @@
+package eu.mikroskeem.firefly.misc;
+
+import lombok.RequiredArgsConstructor;
+import net.md_5.bungee.BungeeCord;
+import net.md_5.bungee.api.config.ConfigurationAdapter;
+import net.md_5.bungee.api.config.ListenerInfo;
+import net.md_5.bungee.api.config.ServerInfo;
+
+import java.util.Collection;
+import java.util.Map;
+
+/**
+ * @author Mark Vainomaa
+ */
+@RequiredArgsConstructor
+public final class ResourceBundleConfigurationWrapper implements ConfigurationAdapter {
+    private final BungeeCord bungeeCord;
+
+    @Override
+    public void load() {}
+
+    @Override
+    public String getString(String path, String def) {
+        return bungeeCord.getTranslationFromBundle(path);
+    }
+
+    @Override
+    public int getInt(String path, int def) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public boolean getBoolean(String path, boolean def) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Collection<?> getList(String path, Collection<?> def) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Map<String, ServerInfo> getServers() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Collection<ListenerInfo> getListeners() {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Collection<String> getGroups(String player) {
+        throw new UnsupportedOperationException();
+    }
+
+    @Override
+    public Collection<String> getPermissions(String group) {
+        throw new UnsupportedOperationException();
+    }
+}
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 03bb1da42bf1b9447834decdf9117ca541cc38df..57a35baf121dafa3640d8057482c3627b409a741 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -113,6 +113,7 @@ public class BungeeCord extends ProxyServer
      */
     private ResourceBundle baseBundle;
     private ResourceBundle customBundle;
+    private ConfigurationAdapter localizationBundleWrapper; // Firefly - Wrap localization into configuration API
     public EventLoopGroup bossEventLoopGroup, workerEventLoopGroup;
     /**
      * locations.yml save thread.
@@ -194,6 +195,8 @@ public class BungeeCord extends ProxyServer
         switch (configurationType) {
             case BUNGEECORD:
                 return configurationAdapter;
+            case BUNGEECORD_MESSAGES:
+                return localizationBundleWrapper;
             case WATERFALL:
                 return waterfallConfigurationAdapter;
             // Firefly start - Fork specific configuration
@@ -210,6 +213,9 @@ public class BungeeCord extends ProxyServer
             case BUNGEECORD:
                 configurationAdapter = adapter;
                 break;
+            case BUNGEECORD_MESSAGES:
+                localizationBundleWrapper = adapter;
+                break;
             case WATERFALL:
                 waterfallConfigurationAdapter = adapter;
                 break;
@@ -244,6 +250,7 @@ public class BungeeCord extends ProxyServer
         {
             baseBundle = ResourceBundle.getBundle( "messages", Locale.ENGLISH );
         }
+        localizationBundleWrapper = new eu.mikroskeem.firefly.misc.ResourceBundleConfigurationWrapper(this); // Firefly - Wrap localization into configuration API
         reloadMessages();
 
         // This is a workaround for quite possibly the weirdest bug I have ever encountered in my life!
@@ -598,6 +605,7 @@ public class BungeeCord extends ProxyServer
 
     public void reloadMessages()
     {
+        localizationBundleWrapper.load(); // Firefly - Wrap localization into configuration API
         File file = new File( "messages.properties" );
         if ( file.isFile() )
         {
@@ -613,6 +621,16 @@ public class BungeeCord extends ProxyServer
 
     @Override
     public String getTranslation(String name, Object... args)
+    // Firefly start - Wrap localization into configuration API
+    {
+        String translation = "<translation '" + name + "' missing>";
+        String string;
+        if((string = this.localizationBundleWrapper.getString(name, (String) null)) != null)
+            translation = (args.length == 0) ? string : MessageFormat.format(string, args);
+        return translation;
+    }
+    public String getTranslationFromBundle(String name, Object... args)
+    // Firefly end
     {
         String translation = "<translation '" + name + "' missing>";
         try
