From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sat, 13 Apr 2019 23:10:17 +0300
Subject: [PATCH] Set sane default configuration values


diff --git a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
index 79f15263bb97329fa5cf10242417b40aa1118553..26a71e76fa996ac3b472f441b0b34be179a9b28c 100644
--- a/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
+++ b/proxy/src/main/java/io/github/waterfallmc/waterfall/conf/WaterfallConfiguration.java
@@ -16,7 +16,7 @@ public abstract class WaterfallConfiguration extends Configuration { // MikroCor
      * <p>
      * Default is true
      */
-    private boolean logInitialHandlerConnections = true;
+    private boolean logInitialHandlerConnections = false; // MikroCord - sane configuration defaults
 
     /**
      * The supported versions displayed to the client
diff --git a/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java b/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
index 51ddb4121454aa8589b8cfddfcc73073882b7d4f..5c8901f9c722f61c7b5e47c44b2eebba712ab180 100644
--- a/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
+++ b/proxy/src/main/java/net/md_5/bungee/conf/Configuration.java
@@ -39,7 +39,7 @@ public abstract class Configuration implements ProxyConfig
     /**
      * UUID used for metrics.
      */
-    private String uuid = UUID.randomUUID().toString();
+    private String uuid; /*= UUID.randomUUID().toString();*/ // MikroCord - sane configuration defaults
     /**
      * Set of all listeners.
      */
@@ -57,12 +57,12 @@ public abstract class Configuration implements ProxyConfig
      * Whether we log proxy commands to the proxy log
      */
     private boolean logCommands;
-    private boolean logPings = true;
+    private boolean logPings; // MikroCord - sane configuration defaults
     private int playerLimit = -1;
     private Collection<String> disabledCommands;
     private int throttle = 4000;
     private int throttleLimit = 3;
-    private boolean ipForward;
+    private boolean ipForward = true; // MikroCord - sane configuration defaults
     private Favicon favicon;
     private int compressionThreshold = 256;
     private boolean preventProxyConnections;
@@ -88,7 +88,7 @@ public abstract class Configuration implements ProxyConfig
 
         listeners = adapter.getListeners();
         timeout = adapter.getInt( "timeout", timeout );
-        uuid = adapter.getString( "stats", uuid );
+        uuid = adapter.getString( "stats", () -> UUID.randomUUID().toString() ); // MikroCord - sane configuration defaults
         onlineMode = adapter.getBoolean( "online_mode", onlineMode );
         logCommands = adapter.getBoolean( "log_commands", logCommands );
         logPings = adapter.getBoolean( "log_pings", logPings );
diff --git a/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java b/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
index 5cd8494017a6d302f2b900504352feef27a1f56d..e6b44d652d53ade95612b7e4d0f2dc71514e89a9 100644
--- a/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
+++ b/proxy/src/main/java/net/md_5/bungee/conf/YamlConfig.java
@@ -110,7 +110,11 @@ public class YamlConfig implements ConfigurationAdapter
         Map<String, Object> groups = get( "groups", null );
         if ( groups == null )
         {
+            /* // MikroCord start - sane configuration defaults
             set( "groups.md_5", Collections.singletonList( "admin" ) );
+            */
+            set("groups", Collections.emptyMap()); // MikroCord - sane configuration defaults
+            // MikroCord end
         }
     }
 
@@ -244,7 +248,7 @@ public class YamlConfig implements ConfigurationAdapter
             new HashMap()
         } ) );
         Map<String, String> forcedDef = new HashMap<>();
-        forcedDef.put( "pvp.md-5.net", "pvp" );
+        //forcedDef.put( "pvp.md-5.net", "pvp" ); MikroCord - sane configuration defaults
 
         Collection<ListenerInfo> ret = new HashSet<>();
 
