From 27766b0d327606b38496e6d5576a89227a5df7b0 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sun, 14 Apr 2019 17:18:40 +0300
Subject: [PATCH] Fix potential race condition


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index b55cb83d..4c96c5dd 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -604,7 +604,14 @@ public class BungeeCord extends ProxyServer
     @Override
     public int getOnlineCount()
     {
+        // MikroCord start - fix potential race condition
+        connectionLock.readLock().lock();
+        try {
         return connections.size();
+        } finally {
+            connectionLock.readLock().unlock();
+        }
+        // MikroCord end
     }
 
     @Override
-- 
2.21.0

