From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Fri, 19 Apr 2019 17:44:34 +0300
Subject: [PATCH] Copy new messages.properties if not present


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index fe2256f18064aee48afd173f329f6f1329a1df2e..699f40a5643ef9974115d199890199ee40ebe73c 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -623,6 +623,15 @@ public class BungeeCord extends ProxyServer
                 getLogger().log( Level.SEVERE, "Could not load custom messages.properties", ex );
             }
         }
+        // Firefly start - copy new messages.properties if not present
+        else if(localizationBundleWrapper instanceof eu.mikroskeem.firefly.misc.ResourceBundleConfigurationWrapper) {
+            try(java.io.InputStream stream = BungeeCord.class.getResourceAsStream("/messages.properties")) {
+                java.nio.file.Files.copy(stream, file.toPath());
+            } catch (IOException e) {
+                getLogger().log(Level.WARNING, "Could not copy default messages.properties", e);
+            }
+        }
+        // Firefly end
     }
 
     @Override
