From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Fri, 19 Apr 2019 18:31:13 +0300
Subject: [PATCH] Add option to ignore certain commands from logging


diff --git a/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java b/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
index 3507e4b235fef56dc1ce7c23a9412b723558f54e..6b695652c71740d3fb3e55651b1a913a53011fa7 100644
--- a/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
+++ b/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
@@ -2,6 +2,8 @@ package eu.mikroskeem.firefly.api.config;
 
 import org.checkerframework.checker.nullness.qual.NonNull;
 
+import java.util.Set;
+
 /**
  * @author Mark Vainomaa
  */
@@ -17,4 +19,10 @@ public interface FireflyProxyConfig {
      */
     @NonNull
     String getRestartScriptPath();
+
+    /**
+     * @return Commands excluded from logging
+     */
+    @NonNull
+    Set<@NonNull String> getIgnoredLogCommands();
 }
diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
index b714503bef028b3fca0e72701477d87d8c28690d..e8447d755452609d796f3ef02be2f698b5471fce 100644
--- a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
@@ -207,6 +207,7 @@ public final class PluginManager
         try { // Waterfall - split command & tab complete exception handlers for exception event
                 if ( proxy.getConfig().isLogCommands() )
                 {
+                    if(!proxy.getConfig().getIgnoredLogCommands().contains(split[0].toLowerCase(Locale.ROOT))) // Firefly - Add option to exclude certain commands from logging
                     proxy.getLogger().log( Level.INFO, "{0} executed command: /{1}", new Object[]
                     {
                         sender.getName(), commandLine
diff --git a/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java b/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
index 1692f8efe71f8e3ac6ae154af46a6d5f8a36f9cc..a0efffca96f0d9fd5be3138630bbecbdce37569a 100644
--- a/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
+++ b/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
@@ -1,10 +1,14 @@
 package eu.mikroskeem.firefly.conf;
 
+import com.google.common.collect.ImmutableList;
 import io.github.waterfallmc.waterfall.conf.WaterfallConfiguration;
 import lombok.Getter;
 import net.md_5.bungee.api.ProxyServer;
 import net.md_5.bungee.api.config.ConfigurationAdapter;
 
+import java.util.HashSet;
+import java.util.Set;
+
 /**
  * @author Mark Vainomaa
  */
@@ -15,6 +19,9 @@ public class FireflyConfiguration extends WaterfallConfiguration {
     @Getter
     private String restartScriptPath = "./restart.sh";
 
+    @Getter
+    private Set<String> ignoredLogCommands;
+
     @Override
     public void load() {
         super.load();
@@ -23,5 +30,6 @@ public class FireflyConfiguration extends WaterfallConfiguration {
 
         restartMessage = config.getString("restart.message", restartMessage);
         restartScriptPath = config.getString("restart.script_path", restartScriptPath);
+        ignoredLogCommands = new HashSet<>(config.getList("log.ignored_commands", () -> ImmutableList.of("login", "register", "changepassword")));
     }
 }
