From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Fri, 19 Apr 2019 17:44:34 +0300
Subject: [PATCH] Copy new messages.properties if not present


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index c984027d422fab22182886996c2b39462f3dae16..2187c955e78ea8d88156a6203a4f6675b691812c 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -578,6 +578,15 @@ public class BungeeCord extends ProxyServer
                 getLogger().log( Level.SEVERE, "Could not load custom messages.properties", ex );
             }
         }
+        // MikroCord start - copy new messages.properties if not present
+        else if(localizationBundleWrapper instanceof eu.mikroskeem.mikrocord.misc.ResourceBundleConfigurationWrapper) {
+            try(java.io.InputStream stream = BungeeCord.class.getResourceAsStream("/messages.properties")) {
+                java.nio.file.Files.copy(stream, file.toPath());
+            } catch (IOException e) {
+                getLogger().log(Level.WARNING, "Could not copy default messages.properties", e);
+            }
+        }
+        // MikroCord end
     }
 
     @Override
