From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Wed, 29 May 2019 20:40:02 +0300
Subject: [PATCH] Speed up shutdown


diff --git a/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java b/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
index d228ec40c5096a1b18ac1fc26d179781cdb1fb25..461b2641ec2445927130e258379c3235eb070c5d 100644
--- a/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
+++ b/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
@@ -52,4 +52,9 @@ public interface FireflyProxyConfig {
      */
     @NonNull
     TCPFastOpenMode getTcpFastOpenMode();
+
+    /**
+     * @return Whether proxy should wait for all players to disconnect infinitely or not
+     */
+    boolean isWaitForAllPlayersToDisconnectInfinitely();
 }
diff --git a/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java b/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
index 2d5d57bdafa876053a0d17fa9517a03fab93adee..374d68b9ea690a864dec33b9772dbe713a688618 100644
--- a/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
+++ b/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
@@ -1,8 +1,8 @@
 package eu.mikroskeem.firefly.conf;
 
 import com.google.common.collect.ImmutableList;
-import eu.mikroskeem.firefly.api.network.TCPFastOpenMode;
 import eu.mikroskeem.firefly.api.network.EventLoopGroupType;
+import eu.mikroskeem.firefly.api.network.TCPFastOpenMode;
 import eu.mikroskeem.firefly.misc.VelocitySupport;
 import io.github.waterfallmc.waterfall.conf.WaterfallConfiguration;
 import lombok.Getter;
@@ -43,6 +43,9 @@ public class FireflyConfiguration extends WaterfallConfiguration {
     @Getter
     private TCPFastOpenMode tcpFastOpenMode = TCPFastOpenMode.CLIENT_ONLY;
 
+    @Getter
+    private boolean waitForAllPlayersToDisconnectInfinitely = true;
+
     @Override
     public void load() {
         super.load();
@@ -57,6 +60,7 @@ public class FireflyConfiguration extends WaterfallConfiguration {
         velocityForwardingSecret = config.getString("velocity_modern_forwarding.secret", () -> VelocitySupport.generateRandomString(12)).getBytes(StandardCharsets.UTF_8);
         tcpFastOpenEnabled = config.getBoolean("networking.tcp_fast_open.enabled", tcpFastOpenEnabled);
         tcpFastOpenMode = setupTfo(config.getInt("networking.tcp_fast_open.mode", tcpFastOpenMode.value));
+        waitForAllPlayersToDisconnectInfinitely = config.getBoolean("shutdown.wait_for_all_players_to_disconnect_infinitely", waitForAllPlayersToDisconnectInfinitely);
     }
 
     private TCPFastOpenMode setupTfo(int value) {
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index c9972507556714e6be8070070ffa90170c1d488f..de20a48e80913f4505d135144144971693b7ebf9 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -496,9 +496,11 @@ public class BungeeCord extends ProxyServer
         stopListeners();
         getLogger().info( "Closing pending connections" );
 
+        Collection<UserConnection> disconnectingPlayers; // Firefly - wait for all players to disconnect instead of using hardcoded delay
         connectionLock.readLock().lock();
         try
         {
+            disconnectingPlayers = java.util.Set.copyOf(connections.values()); // Firefly - wait for all players to disconnect instead of using hardcoded delay
             getLogger().log( Level.INFO, "Disconnecting {0} connections", connections.size() );
             for ( UserConnection user : connections.values() )
             {
@@ -509,12 +511,32 @@ public class BungeeCord extends ProxyServer
             connectionLock.readLock().unlock();
         }
 
+        /* Firefly start - wait for all players to disconnect instead of using hardcoded delay
         try
         {
             Thread.sleep( 500 );
         } catch ( InterruptedException ex )
         {
         }
+        */
+        try {
+            var disconnectFuture = java.util.concurrent.CompletableFuture
+                    .allOf(disconnectingPlayers.stream()
+                            .map(UserConnection::getDisconnectFuture)
+                            .toArray(java.util.concurrent.CompletableFuture[]::new));
+            if (config.isWaitForAllPlayersToDisconnectInfinitely()) {
+                disconnectFuture.get();
+            } else {
+                disconnectFuture.get(10, TimeUnit.SECONDS);
+            }
+        } catch (java.util.concurrent.ExecutionException e) {
+            throw new RuntimeException(e);
+        } catch (java.util.concurrent.TimeoutException e) {
+            getLogger().log(Level.WARNING, "Timeout while waiting all players to clean up - something is taking too long!", e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+        }
+        // Firefly end
 
         if ( reconnectHandler != null )
         {
@@ -548,8 +570,8 @@ public class BungeeCord extends ProxyServer
         }
 
         getLogger().info( "Closing IO threads" );
-                bossEventLoopGroup.shutdownGracefully();
-                workerEventLoopGroup.shutdownGracefully();
+                bossEventLoopGroup.shutdownGracefully(100, 5000, TimeUnit.MILLISECONDS); // Firefly - speed up shutdown
+                workerEventLoopGroup.shutdownGracefully(100, 5000, TimeUnit.MILLISECONDS); // Firefly - speed up shutdown
                 while (true) {
                     try {
                         bossEventLoopGroup.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 47e9b9c38efe945b8f4bc86a0d3fbeb8c4cd8700..ad2b38d287322b6eb0e698cb10e6ebb8d9c17afe 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -152,6 +152,7 @@ public final class UserConnection implements ProxiedPlayer, eu.mikroskeem.firefl
             ch.write( packet );
         }
     };
+    @Getter private final java.util.concurrent.CompletableFuture<Void> disconnectFuture = new java.util.concurrent.CompletableFuture<>(); // Firefly - wait for all players to disconnect instead of using hardcoded delay
 
     public void init()
     {
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index 4d7b1b23e6394ea6a198a957c91e66596d567dee..f157f46abfd101eeda17fe16c300ced606a019ea 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -89,6 +89,7 @@ public class UpstreamBridge extends PacketHandler
             }
             con.getServer().disconnect( "Quitting" );
         }
+        this.con.getDisconnectFuture().complete(null); // Firefly - wait for all players to disconnect instead of using hardcoded delay
     }
 
     @Override
