From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Wed, 29 May 2019 20:16:16 +0300
Subject: [PATCH] Improve event loop group implementation picking


diff --git a/api/src/main/java/eu/mikroskeem/mikrocord/api/network/EventLoopGroupType.java b/api/src/main/java/eu/mikroskeem/mikrocord/api/network/EventLoopGroupType.java
new file mode 100644
index 0000000000000000000000000000000000000000..5ba35f16b22aabe70a9df611ffbe8310d1445a12
--- /dev/null
+++ b/api/src/main/java/eu/mikroskeem/mikrocord/api/network/EventLoopGroupType.java
@@ -0,0 +1,9 @@
+package eu.mikroskeem.mikrocord.api.network;
+
+/**
+ * @author Mark Vainomaa
+ */
+public enum EventLoopGroupType {
+    EPOLL,
+    NIO
+}
\ No newline at end of file
diff --git a/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
index 1df8934ab767a27f1e8a4e49816d79200024ece7..3380011f0137b6f0974a4675c156e944cfc8bd15 100644
--- a/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
+++ b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
@@ -2,11 +2,13 @@ package eu.mikroskeem.mikrocord.conf;
 
 import com.google.common.collect.ImmutableList;
 import eu.mikroskeem.mikrocord.api.network.TCPFastOpenMode;
+import eu.mikroskeem.mikrocord.api.network.EventLoopGroupType;
 import eu.mikroskeem.mikrocord.misc.VelocitySupport;
 import io.github.waterfallmc.waterfall.conf.WaterfallConfiguration;
 import lombok.Getter;
 import net.md_5.bungee.api.ProxyServer;
 import net.md_5.bungee.api.config.ConfigurationAdapter;
+import net.md_5.bungee.netty.PipelineUtils;
 
 import java.nio.charset.StandardCharsets;
 import java.util.HashSet;
@@ -58,6 +60,9 @@ public class MikroCordConfiguration extends WaterfallConfiguration {
     }
 
     private TCPFastOpenMode setupTfo(int value) {
+        if (PipelineUtils.getEventLoopGroupType() != EventLoopGroupType.EPOLL) {
+            return TCPFastOpenMode.DISABLED;
+        }
         var actual = TCPFastOpenMode.fromValue(value);
         ProxyServer.getInstance().getLogger().log(Level.INFO, "TCP fast open mode is set to {0}", actual.name());
         return actual;
diff --git a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
index fe03bf507f498581052ab04cf0db8f68330d1093..c962ac439c584594c1a12550e9a185731bb70534 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
@@ -110,12 +110,15 @@ public class PipelineUtils
     public static final String LEGACY_KICKER = "legacy-kick";
 
     private static boolean epoll;
+    // MikroCord start - Improve event loop group implementation picking
+    @lombok.Getter private static eu.mikroskeem.mikrocord.api.network.EventLoopGroupType eventLoopGroupType;
+    // MikroCord end
 
     static
     {
-        if ( !PlatformDependent.isWindows() && Boolean.parseBoolean( System.getProperty( "bungee.epoll", "true" ) ) )
+        if ( !PlatformDependent.isWindows() && !PlatformDependent.isOsx() && Boolean.parseBoolean( System.getProperty( "bungee.epoll", "true" ) ) ) // MikroCord - Improve event loop group implementation picking
         {
-            ProxyServer.getInstance().getLogger().info( "Not on Windows, attempting to use enhanced EpollEventLoop" );
+            ProxyServer.getInstance().getLogger().info("Attempting to use enhanced EpollEventLoop"); //"Not on Windows, attempting to use enhanced EpollEventLoop" ); // MikroCord start - Improve event loop group implementation picking
 
             if ( epoll = Epoll.isAvailable() )
             {
@@ -125,6 +128,7 @@ public class PipelineUtils
                 ProxyServer.getInstance().getLogger().log( Level.WARNING, "Epoll is not working, falling back to NIO: {0}", Util.exception( Epoll.unavailabilityCause() ) );
             }
         }
+        eventLoopGroupType = epoll ? eu.mikroskeem.mikrocord.api.network.EventLoopGroupType.EPOLL : eu.mikroskeem.mikrocord.api.network.EventLoopGroupType.NIO;
     }
 
     public static EventLoopGroup newEventLoopGroup(int threads, ThreadFactory factory)
