From e5397b5db6a0b60d954fa998034ea309f5cb935c Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Fri, 19 Apr 2019 18:31:13 +0300
Subject: [PATCH] Add option to ignore certain commands from logging


diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
index cd2f66a2..67a33110 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
@@ -251,4 +251,10 @@ public interface ProxyConfig
      */
     String getRestartScriptPath();
     // MikroCord end
+    // MikroCord start - Add option to exclude certain commands from logging
+    /**
+     * @return Commands excluded from logging
+     */
+    java.util.Set<String> getIgnoredLogCommands();
+    // MikroCord end
 }
diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
index 4e1989e9..9bede570 100644
--- a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
@@ -199,6 +199,7 @@ public class PluginManager
             {
                 if ( proxy.getConfig().isLogCommands() )
                 {
+                    if(!proxy.getConfig().getIgnoredLogCommands().contains(split[0])) // MikroCord - Add option to exclude certain commands from logging
                     proxy.getLogger().log( Level.INFO, "{0} executed command: /{1}", new Object[]
                     {
                         sender.getName(), commandLine
diff --git a/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
index 633e3093..6f57a3d0 100644
--- a/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
+++ b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
@@ -1,10 +1,15 @@
 package eu.mikroskeem.mikrocord.conf;
 
+import com.google.common.collect.ImmutableSet;
 import io.github.waterfallmc.waterfall.conf.WaterfallConfiguration;
 import lombok.Getter;
 import net.md_5.bungee.api.ProxyServer;
 import net.md_5.bungee.api.config.ConfigurationAdapter;
 
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
 /**
  * @author Mark Vainomaa
  */
@@ -15,6 +20,9 @@ public class MikroCordConfiguration extends WaterfallConfiguration {
     @Getter
     private String restartScriptPath = "./restart.sh";
 
+    @Getter
+    private Set<String> ignoredLogCommands = ImmutableSet.of("login", "register", "changepassword");
+
     @Override
     public void load() {
         super.load();
@@ -23,5 +31,6 @@ public class MikroCordConfiguration extends WaterfallConfiguration {
 
         restartMessage = config.getString("restart.message", restartMessage);
         restartScriptPath = config.getString("restart.script_path", restartScriptPath);
+        ignoredLogCommands = new HashSet<>((List<String>) config.getList("log.ignored_commands", ignoredLogCommands));
     }
 }
-- 
2.21.0

