From 2e11ac6c483ff915d8427920cef7b5c7113fe264 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Wed, 17 Apr 2019 17:51:11 +0300
Subject: [PATCH] Replace spaces in MCPE usernames with underscores


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index 46bf1978..28e6a560 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -345,6 +345,14 @@ public class InitialHandler extends PacketHandler implements PendingConnection
 
                     uniqueId = peHandshake.getUuid();
                     //TODO: loginProfile
+                    // MikroCord start - replace spaces in MCPE usernames with underscores
+                    String peUsername = peHandshake.getUsername();
+                    if(peUsername.indexOf(' ') != -1) {
+                        peUsername = peUsername.replace(' ', '_');
+                        bungee.getLogger().log(Level.INFO, "{0} fixing up MCPE username from ''{1}'' to ''{2}''", new Object[]{ this, peHandshake.getUsername(), peUsername });
+                        peHandshake.setUsername(peUsername);
+                    }
+                    // MikroCord end
                     handle(new LoginRequest(peHandshake.getUsername()));
 
                     if (isOnlineMode() && !peHandshake.isAuthorized()) {
-- 
2.21.0

