From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Wed, 29 May 2019 20:40:02 +0300
Subject: [PATCH] Speed up shutdown


diff --git a/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java b/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java
index ccba966265138d86cacf60ef989db71f17efe893..f7ad5f76e771c50174c182f228487d247daab877 100644
--- a/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java
+++ b/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java
@@ -52,4 +52,9 @@ public interface MikroCordProxyConfig {
      */
     @NonNull
     TCPFastOpenMode getTcpFastOpenMode();
+
+    /**
+     * @return Whether proxy should wait for all players to disconnect infinitely or not
+     */
+    boolean isWaitForAllPlayersToDisconnectInfinitely();
 }
diff --git a/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
index 3380011f0137b6f0974a4675c156e944cfc8bd15..31c6bd3627fc360f3816f2787e0a0373d0b98124 100644
--- a/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
+++ b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
@@ -1,8 +1,8 @@
 package eu.mikroskeem.mikrocord.conf;
 
 import com.google.common.collect.ImmutableList;
-import eu.mikroskeem.mikrocord.api.network.TCPFastOpenMode;
 import eu.mikroskeem.mikrocord.api.network.EventLoopGroupType;
+import eu.mikroskeem.mikrocord.api.network.TCPFastOpenMode;
 import eu.mikroskeem.mikrocord.misc.VelocitySupport;
 import io.github.waterfallmc.waterfall.conf.WaterfallConfiguration;
 import lombok.Getter;
@@ -43,6 +43,9 @@ public class MikroCordConfiguration extends WaterfallConfiguration {
     @Getter
     private TCPFastOpenMode tcpFastOpenMode = TCPFastOpenMode.CLIENT_ONLY;
 
+    @Getter
+    private boolean waitForAllPlayersToDisconnectInfinitely = true;
+
     @Override
     public void load() {
         super.load();
@@ -57,6 +60,7 @@ public class MikroCordConfiguration extends WaterfallConfiguration {
         velocityForwardingSecret = config.getString("velocity_modern_forwarding.secret", () -> VelocitySupport.generateRandomString(12)).getBytes(StandardCharsets.UTF_8);
         tcpFastOpenEnabled = config.getBoolean("networking.tcp_fast_open.enabled", tcpFastOpenEnabled);
         tcpFastOpenMode = setupTfo(config.getInt("networking.tcp_fast_open.mode", tcpFastOpenMode.value));
+        waitForAllPlayersToDisconnectInfinitely = config.getBoolean("shutdown.wait_for_all_players_to_disconnect_infinitely", waitForAllPlayersToDisconnectInfinitely);
     }
 
     private TCPFastOpenMode setupTfo(int value) {
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 60c63cbd8d6805b5dd8b24a6ebadc50cee53d21e..2f14327e3ea864ce3eeb82d9041611e3e5fa856c 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -475,12 +475,39 @@ public class BungeeCord extends ProxyServer
                     connectionLock.readLock().unlock();
                 }
 
+                /* MikroCord start - wait for all players to disconnect instead of using hardcoded delay
                 try
                 {
                     Thread.sleep( 500 );
                 } catch ( InterruptedException ex )
                 {
                 }
+                */
+                Collection<UserConnection> disconnectingPlayers;
+                try {
+                    connectionLock.readLock().lock();
+                    disconnectingPlayers = java.util.Set.copyOf(connections.values());
+                } finally {
+                    connectionLock.readLock().unlock();
+                }
+                try {
+                    var disconnectFuture = java.util.concurrent.CompletableFuture
+                            .allOf(disconnectingPlayers.stream()
+                                    .map(UserConnection::getDisconnectFuture)
+                                    .toArray(java.util.concurrent.CompletableFuture[]::new));
+                    if (config.isWaitForAllPlayersToDisconnectInfinitely()) {
+                        disconnectFuture.get();
+                    } else {
+                        disconnectFuture.get(10, TimeUnit.SECONDS);
+                    }
+                } catch (java.util.concurrent.ExecutionException e) {
+                    throw new RuntimeException(e);
+                } catch (java.util.concurrent.TimeoutException e) {
+                    getLogger().log(Level.WARNING, "Timeout while waiting all players to clean up - something is taking too long!", e);
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+                // MikroCord end
 
                 if ( reconnectHandler != null )
                 {
@@ -511,8 +538,8 @@ public class BungeeCord extends ProxyServer
                 }
 
                 getLogger().info( "Closing IO threads" );
-                bossEventLoopGroup.shutdownGracefully();
-                workerEventLoopGroup.shutdownGracefully();
+                bossEventLoopGroup.shutdownGracefully(100, 5000, TimeUnit.MILLISECONDS); // MikroCord - speed up shutdown
+                workerEventLoopGroup.shutdownGracefully(100, 5000, TimeUnit.MILLISECONDS); // MikroCord - speed up shutdown
                 while (true) {
                     try {
                         bossEventLoopGroup.awaitTermination(Long.MAX_VALUE, TimeUnit.NANOSECONDS);
diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 5d6022071d23197c584271560edf1ba14668c1b4..73e11313ab4130c04576d885158ab7d62854a93e 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -154,6 +154,7 @@ public final class UserConnection implements ProxiedPlayer, eu.mikroskeem.mikroc
             ch.write( packet );
         }
     };
+    @Getter private final java.util.concurrent.CompletableFuture<Void> disconnectFuture = new java.util.concurrent.CompletableFuture<>(); // MikroCord - wait for all players to disconnect instead of using hardcoded delay
 
     public void init()
     {
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
index 51d0c6cfe58e9f59c1c897ecf13a27ca60830fb8..cb52bd781b09cb5c471fdcadb93e70785e934716 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/UpstreamBridge.java
@@ -87,6 +87,7 @@ public class UpstreamBridge extends PacketHandler
             }
             con.getServer().disconnect( "Quitting" );
         }
+        this.con.getDisconnectFuture().complete(null); // MikroCord - wait for all players to disconnect instead of using hardcoded delay
     }
 
     @Override
