From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Thu, 30 May 2019 02:53:08 +0300
Subject: [PATCH] Fix dumb connection message behaviour


diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index c3ed56d5093ac29bfdeb81e404db0069d4c4b0ae..d606c9640aaf6775e8b43ba20a5cadb5d566b804 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -321,6 +321,7 @@ public final class UserConnection implements ProxiedPlayer, eu.mikroskeem.firefl
                 callback.done( ServerConnectRequest.Result.ALREADY_CONNECTED, null );
             }
 
+            else // Firefly - Fix dumb connection message behaviour
             sendMessage( bungee.getTranslation( "already_connected" ) );
             return;
         }
@@ -331,6 +332,7 @@ public final class UserConnection implements ProxiedPlayer, eu.mikroskeem.firefl
                 callback.done( ServerConnectRequest.Result.ALREADY_CONNECTING, null );
             }
 
+            else // Firefly - Fix dumb connection message behaviour
             sendMessage( bungee.getTranslation( "already_connecting" ) );
             return;
         }
