From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Thu, 30 May 2019 02:53:08 +0300
Subject: [PATCH] Fix dumb connection message behaviour


diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 27eb311f54bf4b2a559a12b718eae177d17048c2..edf13103ba64e2ce1da81122ac12303050e7f23e 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -322,6 +322,7 @@ public final class UserConnection implements ProxiedPlayer, eu.mikroskeem.mikroc
                 callback.done( ServerConnectRequest.Result.ALREADY_CONNECTED, null );
             }
 
+            else // MikroCord - Fix dumb connection message behaviour
             sendMessage( bungee.getTranslation( "already_connected" ) );
             return;
         }
@@ -332,6 +333,7 @@ public final class UserConnection implements ProxiedPlayer, eu.mikroskeem.mikroc
                 callback.done( ServerConnectRequest.Result.ALREADY_CONNECTING, null );
             }
 
+            else // MikroCord - Fix dumb connection message behaviour
             sendMessage( bungee.getTranslation( "already_connecting" ) );
             return;
         }
