From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Fri, 7 Jun 2019 21:34:45 +0300
Subject: [PATCH] Check for '*' permission to grant all permissions


diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 363f3519efa81b244dc78c4aadcd92a6e0c93071..696bfa07e8c8ce686012f40d1228ba336b4a46ae 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -573,7 +573,7 @@ public final class UserConnection implements ProxiedPlayer, eu.mikroskeem.firefl
     @Override
     public boolean hasPermission(String permission)
     {
-        return bungee.getPluginManager().callEvent( new PermissionCheckEvent( this, permission, permissions.contains( permission ) ) ).hasPermission();
+        return bungee.getPluginManager().callEvent( new PermissionCheckEvent( this, permission, permissions.contains("*") || permissions.contains( permission ) ) ).hasPermission(); // Firefly - add '*' permission to grant all permissions
     }
 
     @Override
