From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Thu, 30 May 2019 02:53:08 +0300
Subject: [PATCH] Fix dumb connection message behaviour


diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 2a9625a4be2ad513d544eed6a5e7942914232df9..2ed68a19abb4c2a24c7f10efaa2dd3d23f6dc260 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -323,6 +323,7 @@ public final class UserConnection implements ProxiedPlayer, eu.mikroskeem.mikroc
                 callback.done( ServerConnectRequest.Result.ALREADY_CONNECTED, null );
             }
 
+            else // MikroCord - Fix dumb connection message behaviour
             sendMessage( bungee.getTranslation( "already_connected" ) );
             return;
         }
@@ -333,6 +334,7 @@ public final class UserConnection implements ProxiedPlayer, eu.mikroskeem.mikroc
                 callback.done( ServerConnectRequest.Result.ALREADY_CONNECTING, null );
             }
 
+            else // MikroCord - Fix dumb connection message behaviour
             sendMessage( bungee.getTranslation( "already_connecting" ) );
             return;
         }
