From 104fc902a629c44027643bc6dbcd2cc97ef52741 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Thu, 30 May 2019 02:53:08 +0300
Subject: [PATCH] Fix dumb connection message behaviour


diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index adeef5a3..74182dfa 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -329,6 +329,7 @@ public final class UserConnection implements ProxiedPlayer, eu.mikroskeem.mikroc
                 callback.done( ServerConnectRequest.Result.ALREADY_CONNECTED, null );
             }
 
+            else // MikroCord - Fix dumb connection message behaviour
             sendMessage( bungee.getTranslation( "already_connected" ) );
             return;
         }
@@ -339,6 +340,7 @@ public final class UserConnection implements ProxiedPlayer, eu.mikroskeem.mikroc
                 callback.done( ServerConnectRequest.Result.ALREADY_CONNECTING, null );
             }
 
+            else // MikroCord - Fix dumb connection message behaviour
             sendMessage( bungee.getTranslation( "already_connecting" ) );
             return;
         }
-- 
2.22.0

