From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Mon, 22 Jul 2019 10:20:51 +0300
Subject: [PATCH] Add getFrom to ServerSwitchEvent


diff --git a/api/src/main/java/net/md_5/bungee/api/event/ServerSwitchEvent.java b/api/src/main/java/net/md_5/bungee/api/event/ServerSwitchEvent.java
index 4722517aed4f34a214516ebbd3350e1adc1dc4b1..71add23646fd81269733a750f31f2c1f110a534e 100644
--- a/api/src/main/java/net/md_5/bungee/api/event/ServerSwitchEvent.java
+++ b/api/src/main/java/net/md_5/bungee/api/event/ServerSwitchEvent.java
@@ -19,4 +19,11 @@ public class ServerSwitchEvent extends Event
      * Player whom the server is for.
      */
     private final ProxiedPlayer player;
+    // Firefly start - SpigotMC/BungeeCord#2379
+    /**
+     * Server the player is switch from. May be null if is initial proxy
+     * connect.
+     */
+    private final net.md_5.bungee.api.config.ServerInfo from;
+    // Firefly end
 }
diff --git a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
index a2e784fb58267cfbeaf244bfb559f6da27bb1dd7..d87c4d69f732dbb8964261d380f844904e0b2d8c 100644
--- a/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
+++ b/proxy/src/main/java/net/md_5/bungee/ServerConnector.java
@@ -348,10 +348,11 @@ public class ServerConnector extends PacketHandler implements eu.mikroskeem.fire
         user.setServerJoinQueue( null );
         user.setDimensionChange( false );
 
+        net.md_5.bungee.api.config.ServerInfo from = (user.getServer() != null ? user.getServer().getInfo() : null); // Firefly - SpigotMC/BungeeCord#2379
         user.setServer( server );
         ch.getHandle().pipeline().get( HandlerBoss.class ).setHandler( new DownstreamBridge( bungee, user, server ) );
 
-        bungee.getPluginManager().callEvent( new ServerSwitchEvent( user ) );
+        bungee.getPluginManager().callEvent( new ServerSwitchEvent( user, from ) ); // Firefly - SpigotMC/BungeeCord#2379
 
         thisState = State.FINISHED;
 
