From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Fri, 7 Jun 2019 21:34:45 +0300
Subject: [PATCH] Check for '*' permission to grant all permissions


diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index 2ed68a19abb4c2a24c7f10efaa2dd3d23f6dc260..ec9e847c4fc13ead90cf58e30d511bd9ac8e16be 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -551,7 +551,7 @@ public final class UserConnection implements ProxiedPlayer, eu.mikroskeem.mikroc
     @Override
     public boolean hasPermission(String permission)
     {
-        return bungee.getPluginManager().callEvent( new PermissionCheckEvent( this, permission, permissions.contains( permission ) ) ).hasPermission();
+        return bungee.getPluginManager().callEvent( new PermissionCheckEvent( this, permission, permissions.contains("*") || permissions.contains( permission ) ) ).hasPermission(); // MikroCord - add '*' permission to grant all permissions
     }
 
     @Override
