From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Fri, 7 Jun 2019 21:34:45 +0300
Subject: [PATCH] Check for '*' permission to grant all permissions


diff --git a/proxy/src/main/java/net/md_5/bungee/UserConnection.java b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
index bf808488dd64e16596cfa861ff7f92772f6299e5..c5eaac9a6bbce34cd09524fafabf13a06676ce4e 100644
--- a/proxy/src/main/java/net/md_5/bungee/UserConnection.java
+++ b/proxy/src/main/java/net/md_5/bungee/UserConnection.java
@@ -557,7 +557,7 @@ public final class UserConnection implements ProxiedPlayer, eu.mikroskeem.mikroc
     @Override
     public boolean hasPermission(String permission)
     {
-        return bungee.getPluginManager().callEvent( new PermissionCheckEvent( this, permission, permissions.contains( permission ) ) ).hasPermission();
+        return bungee.getPluginManager().callEvent( new PermissionCheckEvent( this, permission, permissions.contains("*") || permissions.contains( permission ) ) ).hasPermission(); // MikroCord - add '*' permission to grant all permissions
     }
 
     @Override
