From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Fri, 27 Sep 2019 11:43:42 +0300
Subject: [PATCH] Bake event handlers after enabling plugins


diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
index f198f40a51084a902d81688e9930ebc058222342..fa9e60e939507887e793e25041945e3af01af548 100644
--- a/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/PluginManager.java
@@ -343,6 +343,7 @@ public final class PluginManager
                 // Waterfall end
             }
         }
+        this.eventBus.doBakeHandlers(); // Firefly - bake event handlers after enabling plugins
     }
 
     private boolean enablePlugin(Map<PluginDescription, Boolean> pluginStatuses, Stack<PluginDescription> dependStack, PluginDescription plugin)
@@ -522,6 +523,11 @@ public final class PluginManager
                     "Listener %s has registered using deprecated subscribe annotation! Please update to @EventHandler.", listener );
         }
         */ // Firefly end
+        // Firefly start - bake event handlers after enabling plugins
+        if (eventBus.isInitialBakeDone()) {
+            ProxyServer.getInstance().getLogger().log(Level.WARNING, "Plugin {0} registered listener {1} post-enable, this is discouraged", new Object[]{ plugin, listener.getClass().getName() });
+        }
+        // Firefly end
         eventBus.register( listener );
         listenersByPlugin.put( plugin, listener );
     }
diff --git a/event/src/main/java/net/md_5/bungee/event/EventBus.java b/event/src/main/java/net/md_5/bungee/event/EventBus.java
index 95d93cdc2b094ee93abea2bf6c71571b6a5ff7c6..e253da9b5af1684ef33d5cd03615f7bc3408829a 100644
--- a/event/src/main/java/net/md_5/bungee/event/EventBus.java
+++ b/event/src/main/java/net/md_5/bungee/event/EventBus.java
@@ -22,10 +22,12 @@ public class EventBus
     private final Map<Class<?>, EventHandlerMethod[]> byEventBaked = new ConcurrentHashMap<>();
     private final Lock lock = new ReentrantLock();
     private final Logger logger;
+    @lombok.Getter private boolean initialBakeDone = false; // Firefly - bake event handlers after enabling plugins
 
     public EventBus()
     {
         this( null );
+        this.initialBakeDone = true; // Firefly - bake event handlers after enabling plugins; This is only used in unit tests
     }
 
     public EventBus(Logger logger)
@@ -125,6 +127,7 @@ public class EventBus
                     currentPriorityMap.put( listener, entry.getValue().toArray( new Method[0] ) );
                     // Firefly end
                 }
+                if (initialBakeDone) // Firefly - bake event handlers after enabling plugins
                 bakeHandlers( e.getKey() );
             }
         } finally
@@ -161,6 +164,7 @@ public class EventBus
                         byListenerAndPriority.remove( e.getKey() );
                     }
                 }
+                if (initialBakeDone) // Firefly - bake event handlers after enabling plugins
                 bakeHandlers( e.getKey() );
             }
         } finally
@@ -169,6 +173,27 @@ public class EventBus
         }
     }
 
+    // Firefly start - bake event handlers after enabling plugins
+    public void doBakeHandlers() {
+        if (initialBakeDone) {
+            return;
+        }
+
+        try {
+            lock.lock();
+
+            var eventClasses = byListenerAndPriority.keySet();
+            for (Class<?> eventClass : eventClasses) {
+                bakeHandlers(eventClass);
+            }
+        } finally {
+            lock.unlock();
+        }
+
+        initialBakeDone = true;
+    }
+    // Firefly end
+
     /**
      * Shouldn't be called without first locking the writeLock; intended for use
      * only inside {@link #register(java.lang.Object) register(Object)} or
