From 52755c7a184319a558ccef3c28bf39e311b4759e Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sun, 9 Jun 2019 01:47:21 +0300
Subject: [PATCH] Update dependencies


diff --git a/chat/pom.xml b/chat/pom.xml
index de9fefd8..b0e0c841 100644
--- a/chat/pom.xml
+++ b/chat/pom.xml
@@ -22,7 +22,7 @@
         <dependency>
             <groupId>com.google.code.gson</groupId>
             <artifactId>gson</artifactId>
-            <version>2.8.0</version>
+            <version>2.8.5</version> <!-- MikroCord - update dependencies -->
             <scope>compile</scope>
         </dependency>
         <!-- MikroCord start - add KyoriPowered/text support -->
diff --git a/pom.xml b/pom.xml
index b58f38a8..2a855792 100644
--- a/pom.xml
+++ b/pom.xml
@@ -100,7 +100,7 @@
     <properties>
         <maven.deploy.skip>true</maven.deploy.skip> <!-- Waterfall -->
         <build.number>unknown</build.number>
-        <netty.version>4.1.34.Final</netty.version>
+        <netty.version>4.1.36.Final</netty.version> <!-- MikroCord - update dependencies -->
         <!-- Require Java 8 -->
         <maven.compiler.source>11</maven.compiler.source> <!-- MikroCord - Require Java 11 -->
         <maven.compiler.target>11</maven.compiler.target> <!-- MikroCord - Require Java 11 -->
@@ -117,7 +117,7 @@
         <dependency>
             <groupId>com.google.guava</groupId>
             <artifactId>guava</artifactId>
-            <version>20.0</version>
+            <version>28.0-jre</version> <!-- MikroCord - update dependencies -->
             <scope>compile</scope>
         </dependency>
         <dependency>
diff --git a/proxy/pom.xml b/proxy/pom.xml
index c4ec4145..f4b287cb 100644
--- a/proxy/pom.xml
+++ b/proxy/pom.xml
@@ -102,7 +102,7 @@
         <dependency>
             <groupId>mysql</groupId>
             <artifactId>mysql-connector-java</artifactId>
-            <version>5.1.47</version>
+            <version>8.0.16</version> <!-- MikroCord - update dependencies -->
             <scope>runtime</scope>
         </dependency>
 
@@ -179,8 +179,16 @@
                     <artifactSet>
                         <excludes>
                             <exclude>com.google.guava:guava</exclude>
+                            <exclude>com.google.guava:failureaccess</exclude>
+                            <exclude>com.google.guava:listenablefuture</exclude>
+                            <exclude>com.google.code.findbugs:jsr305</exclude>
+                            <exclude>com.google.errorprone:error_prone_annotations</exclude>
+                            <exclude>com.google.j2objc:j2objc-annotations</exclude>
+                            <exclude>org.codehaus.mojo:animal-sniffer-annotations</exclude>
                             <exclude>mysql:mysql-connector-java</exclude>
+                            <exclude>com.google.protobuf:protobuf-java</exclude>
                             <exclude>net.sf.trove4j:core</exclude>
+                            <exclude>net.sf.jopt-simple:jopt-simple</exclude>
                             <exclude>org.yaml:snakeyaml</exclude>
                             <exclude>com.google.code.gson:gson</exclude>
                             <exclude>io.netty:*</exclude>
diff --git a/proxy/src/main/java/eu/mikroskeem/mikrocord/dependencies/DependencyClassLoader.java b/proxy/src/main/java/eu/mikroskeem/mikrocord/dependencies/DependencyClassLoader.java
index 1ca62370..26fb4900 100644
--- a/proxy/src/main/java/eu/mikroskeem/mikrocord/dependencies/DependencyClassLoader.java
+++ b/proxy/src/main/java/eu/mikroskeem/mikrocord/dependencies/DependencyClassLoader.java
@@ -40,34 +40,37 @@ public final class DependencyClassLoader extends URLClassLoader {
             "https://repo.maven.apache.org/maven2"
     );
 
-    private static final String NETTY_VERSION = "4.1.34.Final";
+    private static final String NETTY_VERSION = "4.1.36.Final";
 
     private static final List<Dependency> DEPENDENCIES = Arrays.asList(
-            new Dependency("com.google.guava", "guava", "20.0", "89507701249388e1ed5ddcf8c41f4ce1be7831ef"),
-            new Dependency("mysql", "mysql-connector-java", "5.1.47", "9de4159aaf2d08817a276610b8114a825fca6cfd"),
+            new Dependency("com.google.guava", "guava", "28.0-jre", "54fed371b4b8a8cce1e94a9abd9620982d3aa54b"),
+            new Dependency("com.google.guava", "failureaccess", "1.0.1", "1dcf1de382a0bf95a3d8b0849546c88bac1292c9"),
+            new Dependency("mysql", "mysql-connector-java", "8.0.17", "53da6afdc5f7b45caaa5bc3627028b2041a36cee"),
+            new Dependency("com.google.protobuf", "protobuf-java", "3.6.1", "0d06d46ecfd92ec6d0f3b423b4cd81cb38d8b924"),
             new Dependency("net.sf.trove4j", "core", "3.1.0", "5fd0207b685536b29dd65c86f4b3bf89befb885c"),
+            new Dependency("net.sf.jopt-simple", "jopt-simple", "4.9", "ee9e9eaa0a35360dcfeac129ff4923215fd65904"),
             new Dependency("org.yaml", "snakeyaml", "1.23", "ec62d74fe50689c28c0ff5b35d3aebcaa8b5be68"),
-            new Dependency("com.google.code.gson", "gson", "2.8.0", "c4ba5371a29ac9b2ad6129b1d39ea38750043eff"),
+            new Dependency("com.google.code.gson", "gson", "2.8.5", "f645ed69d595b24d4cf8b3fbb64cc505bede8829"),
             new Dependency("com.flowpowered", "flow-nbt", "1.0.0", "cdb46ca1a493dfc635e53a5055ed6c7872c32941"),
-            new Dependency("io.netty", "netty-buffer", NETTY_VERSION, "8afc2027eadc7848127553ebb490ea0e2b199d4e"),
-            new Dependency("io.netty", "netty-codec", NETTY_VERSION, "0cb53baf46cc55768f2e28d15cde399063555ea8"),
-            new Dependency("io.netty", "netty-codec-dns", NETTY_VERSION, "40352df88786faf68901fa01b0c51baafcad5115"),
-            new Dependency("io.netty", "netty-codec-haproxy", NETTY_VERSION, "d8f64c174e7875e3be6a75f91b0ee3ba3316655e"),
-            new Dependency("io.netty", "netty-codec-http", NETTY_VERSION, "02887d87fbc1b057657348f61dc538f7296daf79"),
-            new Dependency("io.netty", "netty-common", NETTY_VERSION, "2dffa21967d36cae446b6905bd5fb39750fcba43"),
-            new Dependency("io.netty", "netty-handler", NETTY_VERSION, "96169098bebda3d1432f5421c338958063b3da52"),
-            new Dependency("io.netty", "netty-resolver", NETTY_VERSION, "528789e98eef519960a56e44569ecb2260782fd7"),
-            new Dependency("io.netty", "netty-resolver-dns", NETTY_VERSION, "0ca536138bcc44997354cdd102933656f65f56c7"),
-            new Dependency("io.netty", "netty-transport-native-epoll", NETTY_VERSION, "linux-x86_64", "4ee3e61c2429f280a63729d39b3973562318600b"),
-            new Dependency("io.netty", "netty-transport-native-unix-common", NETTY_VERSION, "7e63ead8935a85d601646d44263a40fabe225771"),
-            new Dependency("io.netty", "netty-transport", NETTY_VERSION, "f57be6509efa6766d90e537a775930101f26b35c"),
+            new Dependency("io.netty", "netty-buffer", NETTY_VERSION, "7f2db0921dd57df4db076229830ab09bba713aeb"),
+            new Dependency("io.netty", "netty-codec", NETTY_VERSION, "8462116d327bb3d1ec24258071f2e7345a73dbfc"),
+            new Dependency("io.netty", "netty-codec-dns", NETTY_VERSION, "8775964a38d6e5d2847c98fc91a91486665b1dcc"),
+            new Dependency("io.netty", "netty-codec-haproxy", NETTY_VERSION, "20f23c1e5a374cd04d76bb480ab3418f5775f7b2"),
+            new Dependency("io.netty", "netty-codec-http", NETTY_VERSION, "62b73d439dbddf3c0dde092b048580139695ab46"),
+            new Dependency("io.netty", "netty-common", NETTY_VERSION, "f6f38fde652a70ea579897edc80e52353e487ae6"),
+            new Dependency("io.netty", "netty-handler", NETTY_VERSION, "1c38a5920a10c01b1cce4cdc964447ec76abf1b5"),
+            new Dependency("io.netty", "netty-resolver", NETTY_VERSION, "e4d243fbf4e6837fa294f892bf97149e18129100"),
+            new Dependency("io.netty", "netty-resolver-dns", NETTY_VERSION, "071f0d061465651d90c7f1d40a66580f69821582"),
+            new Dependency("io.netty", "netty-transport-native-epoll", NETTY_VERSION, "linux-x86_64", "8ec8e248ef34b86cc7fae023fb40b9ffa8f5614f"),
+            new Dependency("io.netty", "netty-transport-native-unix-common", NETTY_VERSION, "d95d7033f400f9472db9da7834c443b96cd4bab0"),
+            new Dependency("io.netty", "netty-transport", NETTY_VERSION, "8546e6be47be587acab86bbd106ca023678f07d9"),
             new Dependency("org.apache.logging.log4j", "log4j-iostreams", "2.11.1", "54219892aca8fe8b91d5c0e4e74d78b5ea613b1d"),
             new Dependency("org.apache.logging.log4j", "log4j-jul", "2.11.1", "f94ff1933609c93844d48815eecbe9e7e4d26473"),
             new Dependency("org.apache.logging.log4j", "log4j-slf4j-impl", "2.11.1", "4b41b53a3a2d299ce381a69d165381ca19f62912"),
             new Dependency("org.slf4j", "slf4j-api", "1.7.25", "da76ca59f6a57ee3102f8f9bd9cee742973efa8a"),
-            new Dependency("net.kyori", "text-api", "3.0.1", "664de5e3f01adad302adb0bd1d7933203463caba"),
-            new Dependency("net.kyori", "text-serializer-gson", "3.0.1", "e3c395e7ed613b48f6d4963c5127307b3e3666bb"),
-            new Dependency("net.kyori", "text-serializer-legacy", "3.0.1", "6fdbf35ca215180bd65a9a8805374a4c35b86cff"),
+            new Dependency("net.kyori", "text-api", "3.0.2", "608cdb44a74bbd68745941760df730ed55e4b47c"),
+            new Dependency("net.kyori", "text-serializer-gson", "3.0.2", "9ac22f04f3504c52ff1618c5a8d9a6145d8d9c9e"),
+            new Dependency("net.kyori", "text-serializer-legacy", "3.0.2", "8acbfb36356259273a8e3a15782e4f2980375bc5"),
             new Dependency("org.ow2.asm", "asm", "7.1", "fa29aa438674ff19d5e1386d2c3527a0267f291e")
     );
 
-- 
2.22.0

