From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Fri, 27 Sep 2019 11:12:31 +0300
Subject: [PATCH] Make PluginClassLoader#addURL public


diff --git a/api/src/main/java/net/md_5/bungee/api/plugin/PluginClassloader.java b/api/src/main/java/net/md_5/bungee/api/plugin/PluginClassloader.java
index 141de9e68abc7b5d60a25c3819c4a4cf4985f623..b6adff8d004cd296a5796faf23692c67d93ddb43 100644
--- a/api/src/main/java/net/md_5/bungee/api/plugin/PluginClassloader.java
+++ b/api/src/main/java/net/md_5/bungee/api/plugin/PluginClassloader.java
@@ -59,4 +59,5 @@ public class PluginClassloader extends URLClassLoader implements eu.mikroskeem.m
         return this.defineClass(name, data, 0, data.length);
     }
     // MikroCord end
+    @Override public void addURL(URL url) { super.addURL(url); } // MikroCord - Make PluginClassLoader#addURL public
 }
