From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sat, 20 Jul 2019 16:49:01 +0300
Subject: [PATCH] Allow Query response plugins population


diff --git a/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java b/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java
index 555431dba5926c7aeb287b589a21ba4d74c547d6..25e30d199e0c42949be43ef026099512bba85724 100644
--- a/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java
+++ b/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java
@@ -69,4 +69,9 @@ public interface MikroCordProxyConfig {
      * @return Whether scoreboard team strings should be interned or not
      */
     boolean isInternScoreboardTeamStrings();
+
+    /**
+     * @return Whether plugins list should be populated in GS4 query response or not
+     */
+    boolean isPopulateGS4QueryPlugins();
 }
diff --git a/api/src/main/java/io/github/waterfallmc/waterfall/QueryResult.java b/api/src/main/java/io/github/waterfallmc/waterfall/QueryResult.java
index 9a262c8a11caeade95a084e5e5bc0be398917bbf..eecdcfeb293646f01ab847bec80caf9165618d34 100644
--- a/api/src/main/java/io/github/waterfallmc/waterfall/QueryResult.java
+++ b/api/src/main/java/io/github/waterfallmc/waterfall/QueryResult.java
@@ -56,4 +56,15 @@ public class QueryResult {
      */
     @NonNull
     private String version;
+    // MikroCord start - allow Query response plugins population
+    private final List<PluginInfo> plugins = new java.util.LinkedList<>();
+    @Data
+    public static final class PluginInfo {
+        @NonNull
+        private final String name;
+
+        @NonNull
+        private final String version;
+    }
+    // MikroCord end
 }
diff --git a/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
index daae1811572243f17402b526ade9a04dd9c69e18..4bed1988e841dee904989ddfa24a18164fabc62d 100644
--- a/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
+++ b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
@@ -52,6 +52,9 @@ public class MikroCordConfiguration extends WaterfallConfiguration {
     @Getter
     private boolean internScoreboardTeamStrings = false;
 
+    @Getter
+    private boolean populateGS4QueryPlugins = false;
+
     @Override
     public void load() {
         super.load();
@@ -69,6 +72,7 @@ public class MikroCordConfiguration extends WaterfallConfiguration {
         waitForAllPlayersToDisconnectInfinitely = config.getBoolean("shutdown.wait_for_all_players_to_disconnect_infinitely", waitForAllPlayersToDisconnectInfinitely);
         allowTablistRewrite = config.getBoolean("performance.allow_tablist_rewrite", allowTablistRewrite);
         internScoreboardTeamStrings = config.getBoolean("performance.intern_scoreboard_team_strings", internScoreboardTeamStrings);
+        populateGS4QueryPlugins = config.getBoolean("query.populate_plugins", populateGS4QueryPlugins);
     }
 
     private TCPFastOpenMode setupTfo(int value) {
diff --git a/query/src/main/java/net/md_5/bungee/query/QueryHandler.java b/query/src/main/java/net/md_5/bungee/query/QueryHandler.java
index 257347578f4828d36a02f126222fbc0d1924b74d..115bad2171580594b83bd390dbdbf098f74deb90 100644
--- a/query/src/main/java/net/md_5/bungee/query/QueryHandler.java
+++ b/query/src/main/java/net/md_5/bungee/query/QueryHandler.java
@@ -33,6 +33,7 @@ public class QueryHandler extends SimpleChannelInboundHandler<DatagramPacket>
     private final Random random = new Random();
     private final Cache<InetAddress, QuerySession> sessions = CacheBuilder.newBuilder().expireAfterWrite( 30, TimeUnit.SECONDS ).build();
     private final Cache<InetAddress, Integer> blacklistedAddresses = CacheBuilder.newBuilder().maximumSize(1024).expireAfterWrite( 120, TimeUnit.SECONDS ).build(); // MikroCord - Blacklist invalid query packet spammers
+    private volatile List<QueryResult.PluginInfo> cachedPluginInfoList = null; // MikroCord - allow Query response plugins population
 
     private void writeShort(ByteBuf buf, int s)
     {
@@ -120,12 +121,28 @@ public class QueryHandler extends SimpleChannelInboundHandler<DatagramPacket>
                 throw new IllegalStateException( "No session!" );
             }
 
+            // MikroCord start - allow Query response plugins population
+            List<QueryResult.PluginInfo> plugins = this.cachedPluginInfoList;
+            if (plugins == null) {
+                synchronized (this) {
+                    if (this.cachedPluginInfoList == null) {
+                        this.cachedPluginInfoList = plugins = bungee.getConfig().isPopulateGS4QueryPlugins() ? bungee.getPluginManager().getPlugins()
+                                .stream()
+                                .map(p -> p.getDescription()) // No it cannot be converted, it's going to be uglier here.
+                                .map(p -> new QueryResult.PluginInfo(p.getName(), p.getVersion()))
+                                .collect(Collectors.toList())
+                                : java.util.Collections.emptyList();
+                    }
+                }
+            }
+            // MikroCord end
             // Waterfall start
             List<String> players = bungee.getPlayers().stream().map(ProxiedPlayer::getName).collect(Collectors.toList());
 
             ProxyQueryEvent event = new ProxyQueryEvent(listener, new QueryResult(listener.getMotd(), "SMP", "MikroCord_Proxy",
                     bungee.getOnlineCount(), listener.getMaxPlayers(), listener.getHost().getPort(),
                     listener.getHost().getHostString(), "MINECRAFT",  players, bungee.getGameVersion()));
+            event.getResult().getPlugins().addAll(plugins); // MikroCord - allow Query response plugins population
             QueryResult result = bungee.getPluginManager().callEvent(event).getResult();
             // Waterfall end
 
@@ -159,7 +176,25 @@ public class QueryHandler extends SimpleChannelInboundHandler<DatagramPacket>
                 // Start Extra Info
                 data.put( "game_id", result.getGameId() );
                 data.put( "version", result.getVersion() );
+                /* MikroCord start - allow Query response plugins population
                 data.put( "plugins", "" ); // TODO: Allow population?
+                 */
+                StringBuilder pluginsString = new StringBuilder();
+                pluginsString.append(bungee.getName()).append(' ').append(bungee.getGameVersion());
+                if (!plugins.isEmpty()) {
+                    pluginsString.append(':').append(' ');
+                    java.util.Iterator<QueryResult.PluginInfo> iterator = plugins.iterator();
+                    while (iterator.hasNext()) {
+                        QueryResult.PluginInfo info = iterator.next();
+                        pluginsString.append(info.getName());
+                        pluginsString.append(' ').append(info.getVersion());
+                        if (iterator.hasNext()) {
+                            pluginsString.append(';').append(' ');
+                        }
+                    }
+                }
+                data.put("plugins", pluginsString.toString());
+                // MikroCord end
                 // End Extra Info
                 data.put( "map", result.getWorldName() );
                 data.put( "numplayers", Integer.toString( result.getOnlinePlayers() ) );
