From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sun, 25 Aug 2019 05:00:39 +0300
Subject: [PATCH] Add Prometheus support


diff --git a/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java b/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java
index 52e085e7bdd7ff5845d3c6aab19baf0dc064ad47..a67f3667c40aacba2d7cfe27cc6bd39950f0792a 100644
--- a/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java
+++ b/api/src/main/java/eu/mikroskeem/mikrocord/api/config/MikroCordProxyConfig.java
@@ -79,4 +79,20 @@ public interface MikroCordProxyConfig {
      * @return Whether packet decode exceptions should be noisy or not
      */
     boolean isLogNoisyPacketDecodeException();
+
+    /**
+     * @return Whether Prometheus metrics should be enabled or not
+     */
+    boolean isPrometheusEnabled();
+
+    /**
+     * @return Hostname where Prometheus exporter should listen on
+     */
+    @NonNull
+    String getPrometheusListenHost();
+
+    /**
+     * @return Port where Prometheus exporter should listen on
+     */
+    int getPrometheusListenPort();
 }
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
index 393a23789286ac3c22c1f7790de3ab8f4b7dac06..b0d63201f39c5656afd17daabf07fc61b9a66157 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
@@ -24,6 +24,11 @@ public class MinecraftDecoder extends MessageToMessageDecoder<ByteBuf>
     @Setter private static java.util.function.BooleanSupplier noisyPacketDecodeExceptionSupplier;
     private final boolean noisyPacketDecodeException;
     // MikroCord end
+    // MikroCord start - add Prometheus metrics support
+    @Setter private static Runnable decodedMetric;
+    @Setter private static Runnable emptyPacketMetric;
+    @Setter private static Runnable invalidPacketMetric;
+    // MikroCord end
 
     public MinecraftDecoder(Protocol protocol, boolean server, int protocolVersion) {
         this.protocol = protocol;
@@ -43,6 +48,7 @@ public class MinecraftDecoder extends MessageToMessageDecoder<ByteBuf>
         {
             // Waterfall start
             if (in.readableBytes() == 0) {
+                emptyPacketMetric.run(); // MikroCord - add Prometheus metrics support
                 return;
             }
             // Waterfall end
@@ -60,6 +66,7 @@ public class MinecraftDecoder extends MessageToMessageDecoder<ByteBuf>
                 {
                     throw new BadPacketException( "Did not read all bytes from packet " + packet.getClass() + " " + packetId + " Protocol " + protocol + " Direction " + prot.getDirection() );
                 }
+                decodedMetric.run(); // MikroCord - add Prometheus metrics support
             } else
             {
                 in.skipBytes( in.readableBytes() );
@@ -68,6 +75,7 @@ public class MinecraftDecoder extends MessageToMessageDecoder<ByteBuf>
             out.add( new PacketWrapper( packet, slice ) );
             slice = null;
         } catch (BadPacketException | IndexOutOfBoundsException e) {
+            invalidPacketMetric.run(); // MikroCord - add Prometheus metrics support
             final String packetTypeStr;
             if (packetTypeInfo instanceof Integer) {
                 packetTypeStr = "id " + Integer.toHexString((Integer) packetTypeInfo);
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftEncoder.java b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftEncoder.java
index d4b0384348c17de5c2350c8c5f928fcc0c1868db..073004e479a7707b8a72f7b44974741203d983f0 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftEncoder.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftEncoder.java
@@ -15,6 +15,9 @@ public class MinecraftEncoder extends MessageToByteEncoder<DefinedPacket>
     private boolean server;
     @Setter
     private int protocolVersion;
+    // MikroCord start - add Prometheus metrics support
+    @Setter private static Runnable encodedMetric;
+    // MikroCord end
 
     @Override
     protected void encode(ChannelHandlerContext ctx, DefinedPacket msg, ByteBuf out) throws Exception
@@ -22,5 +25,6 @@ public class MinecraftEncoder extends MessageToByteEncoder<DefinedPacket>
         Protocol.DirectionData prot = ( server ) ? protocol.TO_CLIENT : protocol.TO_SERVER;
         DefinedPacket.writeVarInt( prot.getId( msg.getClass(), protocolVersion ), out );
         msg.write( out, prot.getDirection(), protocolVersion );
+        encodedMetric.run(); // MikroCord - add Prometheus metrics support
     }
 }
diff --git a/proxy/pom.xml b/proxy/pom.xml
index 658766786ec884460927348937e2b94d06169d55..19353857fd7816fafbe6a1d48f4e2aece615e0cb 100644
--- a/proxy/pom.xml
+++ b/proxy/pom.xml
@@ -21,6 +21,7 @@
     <properties>
         <maven.deploy.skip>true</maven.deploy.skip>
         <maven.build.timestamp.format>yyyyMMdd</maven.build.timestamp.format> <!-- MikroCord -->
+        <prometheus.client.version>0.6.0</prometheus.client.version> <!-- MikroCord - add Prometheus metrics support -->
     </properties>
 
     <dependencies>
@@ -119,6 +120,28 @@
             <scope>runtime</scope>
         </dependency>
         <!-- Waterfall end -->
+        <!-- MikroCord start - add Prometheus metrics support -->
+        <dependency>
+            <groupId>io.prometheus</groupId>
+            <artifactId>simpleclient</artifactId>
+            <version>${prometheus.client.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>io.prometheus</groupId>
+            <artifactId>simpleclient_hotspot</artifactId>
+            <version>${prometheus.client.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>io.prometheus</groupId>
+            <artifactId>simpleclient_httpserver</artifactId>
+            <version>${prometheus.client.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>eu.mikroskeem</groupId>
+            <artifactId>jvm-hiccup</artifactId>
+            <version>1.0.0</version>
+        </dependency>
+        <!-- MikroCord end -->
     </dependencies>
 
     <build>
@@ -200,6 +223,8 @@
                             <exclude>net.kyori:*</exclude>
                             <exclude>org.checkerframework:checker-qual</exclude>
                             <exclude>org.ow2.asm:asm</exclude>
+                            <exclude>io.prometheus:*</exclude>
+                            <exclude>eu.mikroskeem:jvm-hiccup</exclude>
                         </excludes>
                     </artifactSet>
                     <!-- MikroCord end -->
diff --git a/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
index 1c796697ff54fefb2b3c86862a19ecc7bd1c63f9..f243777118b19e199411b66b89759a7b85919f45 100644
--- a/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
+++ b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
@@ -64,6 +64,15 @@ public class MikroCordConfiguration extends WaterfallConfiguration {
     @Getter
     private boolean logNoisyPacketDecodeException = false;
 
+    @Getter
+    private boolean prometheusEnabled = true;
+
+    @Getter
+    private String prometheusListenHost = "127.0.0.1";
+
+    @Getter
+    private int prometheusListenPort = 8888;
+
     @Override
     public void load() {
         super.load();
@@ -83,6 +92,9 @@ public class MikroCordConfiguration extends WaterfallConfiguration {
         internScoreboardTeamStrings = config.getBoolean("performance.intern_scoreboard_team_strings", internScoreboardTeamStrings);
         populateGS4QueryPlugins = config.getBoolean("query.populate_plugins", populateGS4QueryPlugins);
         logNoisyPacketDecodeException = config.getBoolean("log.noisy_packet_decode_exception", logNoisyPacketDecodeException);
+        prometheusEnabled = config.getBoolean("prometheus.enabled", prometheusEnabled);
+        prometheusListenHost = config.getString("prometheus.listen_host", prometheusListenHost);
+        prometheusListenPort = config.getInt("prometheus.listen_port", prometheusListenPort);
     }
 
     private TCPFastOpenMode setupTfo(int value) {
diff --git a/proxy/src/main/java/eu/mikroskeem/mikrocord/dependencies/DependencyClassLoader.java b/proxy/src/main/java/eu/mikroskeem/mikrocord/dependencies/DependencyClassLoader.java
index 9d74221aba1ff5c4d126b6db5add22e91cc079d3..824e6616352ac69c9c55ec6961585eb1d3fc5d46 100644
--- a/proxy/src/main/java/eu/mikroskeem/mikrocord/dependencies/DependencyClassLoader.java
+++ b/proxy/src/main/java/eu/mikroskeem/mikrocord/dependencies/DependencyClassLoader.java
@@ -37,10 +37,12 @@ public final class DependencyClassLoader extends URLClassLoader {
     private static final Logger logger = LogManager.getLogger(DependencyClassLoader.class);
     private static final List<String> DEFAULT_REPOSITORIES = Arrays.asList(
             // TODO: mirror repositories?
-            "https://repo.maven.apache.org/maven2"
+            "https://repo.maven.apache.org/maven2",
+            "https://repo.wut.ee/repository/mikroskeem-repo"
     );
 
     private static final String NETTY_VERSION = "4.1.42.Final";
+    private static final String PROMETHEUS_VERSION = "0.6.0";
 
     private static final List<Dependency> DEPENDENCIES = Arrays.asList(
             new Dependency("com.google.guava", "guava", "28.1-jre", "b0e91dcb6a44ffb6221b5027e12a5cb34b841145"),
@@ -71,7 +73,12 @@ public final class DependencyClassLoader extends URLClassLoader {
             new Dependency("net.kyori", "text-api", "3.0.2", "608cdb44a74bbd68745941760df730ed55e4b47c"),
             new Dependency("net.kyori", "text-serializer-gson", "3.0.2", "9ac22f04f3504c52ff1618c5a8d9a6145d8d9c9e"),
             new Dependency("net.kyori", "text-serializer-legacy", "3.0.2", "8acbfb36356259273a8e3a15782e4f2980375bc5"),
-            new Dependency("org.ow2.asm", "asm", "7.1", "fa29aa438674ff19d5e1386d2c3527a0267f291e")
+            new Dependency("org.ow2.asm", "asm", "7.1", "fa29aa438674ff19d5e1386d2c3527a0267f291e"),
+            new Dependency("io.prometheus", "simpleclient", PROMETHEUS_VERSION, "26073e94cbfa6780e10ef524e542cf2a64dabe67"),
+            new Dependency("io.prometheus", "simpleclient_common", PROMETHEUS_VERSION, "08b4f119cfdff67a02a066e6e519bb2bab0a2a1b"),
+            new Dependency("io.prometheus", "simpleclient_hotspot", PROMETHEUS_VERSION, "2703b02c4b2abb078de8365f4ef3b7d5e451382d"),
+            new Dependency("io.prometheus", "simpleclient_httpserver", PROMETHEUS_VERSION, "d46c7273a4dd10e611e3cd617f46eb8b2dae8fdc"),
+            new Dependency("eu.mikroskeem", "jvm-hiccup", "1.0.0", "906515353ac4c529e9fbbb6ad945dddc1dbe7cc6")
     );
 
     private static DependencyClassLoader instance;
diff --git a/proxy/src/main/java/eu/mikroskeem/mikrocord/misc/PrometheusMetrics.java b/proxy/src/main/java/eu/mikroskeem/mikrocord/misc/PrometheusMetrics.java
new file mode 100644
index 0000000000000000000000000000000000000000..14d3775e1c0674d4cf89f90900ebad540661a5ec
--- /dev/null
+++ b/proxy/src/main/java/eu/mikroskeem/mikrocord/misc/PrometheusMetrics.java
@@ -0,0 +1,115 @@
+package eu.mikroskeem.mikrocord.misc;
+
+import eu.mikroskeem.jvmhiccup.HiccupMeterThread;
+import io.prometheus.client.Collector;
+import io.prometheus.client.Counter;
+import io.prometheus.client.Gauge;
+import io.prometheus.client.GaugeMetricFamily;
+import io.prometheus.client.exporter.HTTPServer;
+import io.prometheus.client.hotspot.DefaultExports;
+import net.md_5.bungee.api.ProxyServer;
+import net.md_5.bungee.protocol.MinecraftDecoder;
+import net.md_5.bungee.protocol.MinecraftEncoder;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.function.DoubleSupplier;
+
+/**
+ * @author Mark Vainomaa
+ */
+public final class PrometheusMetrics {
+    private static final Logger logger = LoggerFactory.getLogger(PrometheusMetrics.class);
+    private static final HiccupMeterThread hiccupMeter;
+
+    private static String prevHost = null;
+    private static Integer prevPort = null;
+    private static HTTPServer server = null;
+
+    public static final Gauge JVM_GC_HICCUP = Gauge
+            .build("jvm_gc_hiccup", "JVM GC hiccup")
+            .register();
+
+    public static final Collector CLIENTS_ONLINE = createGauge("clients_online", "Total clients online",
+            () -> ProxyServer.getInstance().getOnlineCount());
+    public static final Collector INSTALLED_PLUGINS = createGauge("plugins_installed", "Total plugins installed",
+            () -> ProxyServer.getInstance().getPluginManager().getPlugins().size());
+    public static final Counter RUNTIME_CONNECTED_CLIENTS = Counter
+            .build("clients_connected", "Total clients connected in proxy lifetime")
+            .register();
+
+    public static final Counter RUNTIME_PACKETS_DECODED = Counter
+            .build("packets_decoded", "Total packets successfully decoded in proxy lifetime")
+            .register();
+
+    public static final Counter RUNTIME_PACKETS_ENCODED = Counter
+            .build("packets_encoded", "Total packets successfully encoded in proxy lifetime")
+            .register();
+
+    public static final Counter RUNTIME_EMPTY_PACKETS_ENCOUNTERED = Counter
+            .build("empty_packets", "Total empty packets encountered in proxy lifetime")
+            .register();
+    public static final Counter RUNTIME_INVALID_PACKETS_ENCOUNTERED = Counter
+            .build("invalid_packets", "Total invalid packets encountered in proxy lifetime")
+            .register();
+
+    private static final boolean ENABLED;
+
+    static {
+        final Runnable noopRunnable = (ENABLED = ProxyServer.getInstance().getConfig().isPrometheusEnabled()) ? null : () -> {};
+        hiccupMeter = new HiccupMeterThread(JVM_GC_HICCUP::set);
+
+        // JVM metrics
+        if (ENABLED) {
+            DefaultExports.initialize();
+            hiccupMeter.start();
+        }
+
+        // Protocol decoder metrics
+        MinecraftDecoder.setDecodedMetric(ENABLED ? RUNTIME_PACKETS_DECODED::inc : noopRunnable);
+        MinecraftDecoder.setEmptyPacketMetric(ENABLED ? RUNTIME_EMPTY_PACKETS_ENCOUNTERED::inc : noopRunnable);
+        MinecraftDecoder.setInvalidPacketMetric(ENABLED ? RUNTIME_INVALID_PACKETS_ENCOUNTERED::inc : noopRunnable);
+
+        // Protocol encoder metrics
+        MinecraftEncoder.setEncodedMetric(ENABLED ? RUNTIME_PACKETS_ENCODED::inc : noopRunnable);
+    }
+
+    public static void initialize() {
+        if (!ENABLED) {
+            return;
+        }
+
+        // Set up exporter server
+        var host = ProxyServer.getInstance().getConfig().getPrometheusListenHost();
+        var port = ProxyServer.getInstance().getConfig().getPrometheusListenPort();
+        if ((prevHost == null || prevPort == null) || (!prevHost.equals(host) || port != prevPort)) {
+            deinitialize();
+            try {
+                server = new HTTPServer(host, port);
+                logger.info("Prometheus HTTP server listening at http://{}:{}", host, port);
+            } catch (IOException e) {
+                logger.warn("Failed to start Prometheus HTTP server at http://{}:{}", host, port, e);
+            }
+        }
+    }
+
+    public static void deinitialize() {
+        if (ENABLED && server != null) {
+            server.stop();
+            logger.info("Prometheus HTTP server stopped");
+        }
+    }
+
+    private static Collector createGauge(String name, String help, DoubleSupplier valueSupplier) {
+        var gauge = new Collector() {
+            @Override
+            public List<MetricFamilySamples> collect() {
+                return Collections.singletonList(new GaugeMetricFamily(name, help, valueSupplier.getAsDouble()));
+            }
+        };
+        return gauge.register();
+    }
+}
diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index 98b29cf195ba04289a151c0f8410b6723834a674..602b58dedb3068fe89611eeb934f8428b75d691d 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -345,6 +345,7 @@ public class BungeeCord extends ProxyServer
         {
             connectionThrottle = new ConnectionThrottle( config.getThrottle(), config.getThrottleLimit() );
         }
+        eu.mikroskeem.mikrocord.misc.PrometheusMetrics.initialize(); // MikroCord - add Prometheus metrics support
         startListeners();
 
         saveThread.scheduleAtFixedRate( new TimerTask()
@@ -461,6 +462,7 @@ public class BungeeCord extends ProxyServer
             public void run()
             {
                 stopListeners();
+                eu.mikroskeem.mikrocord.misc.PrometheusMetrics.deinitialize(); // MikroCord - add Prometheus metrics support
                 getLogger().info( "Closing pending connections" );
 
                 connectionLock.readLock().lock();
@@ -811,6 +813,7 @@ public class BungeeCord extends ProxyServer
             connections.put( con.getName(), con );
             connectionsByUUID.put( con.getUniqueId(), con );
             connectionsByOfflineUUID.put( con.getPendingConnection().getOfflineId(), con );
+            eu.mikroskeem.mikrocord.misc.PrometheusMetrics.RUNTIME_CONNECTED_CLIENTS.inc(); // MikroCord - add Prometheus metrics support
         } finally
         {
             connectionLock.writeLock().unlock();
diff --git a/proxy/src/main/java/net/md_5/bungee/command/CommandReload.java b/proxy/src/main/java/net/md_5/bungee/command/CommandReload.java
index 0ad37d4e3de8dc16cf197d8c96e9fcd70ad7eae5..e91f68fe201ccba4ad9b950aafc27eb677f3ca75 100644
--- a/proxy/src/main/java/net/md_5/bungee/command/CommandReload.java
+++ b/proxy/src/main/java/net/md_5/bungee/command/CommandReload.java
@@ -22,6 +22,7 @@ public class CommandReload extends Command
         BungeeCord.getInstance().stopListeners();
         BungeeCord.getInstance().startListeners();
         BungeeCord.getInstance().getPluginManager().callEvent( new ProxyReloadEvent( sender ) );
+        eu.mikroskeem.mikrocord.misc.PrometheusMetrics.initialize(); // MikroCord - add Prometheus metrics support
 
         sender.sendMessage(ChatColor.GREEN + "MikroCord has been reloaded"); if(true) return; // MikroCord
         sender.sendMessage( ChatColor.BOLD.toString() + ChatColor.RED.toString() + "Waterfall has been reloaded."
