From 20187326b75fd6710c64ec19638a481ce9e0b6df Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Thu, 15 Aug 2019 17:26:50 +0300
Subject: [PATCH] Add option to disable noisy packet decode exception


diff --git a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
index 35b72a4b..aa6e7a72 100644
--- a/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
+++ b/api/src/main/java/net/md_5/bungee/api/ProxyConfig.java
@@ -280,4 +280,7 @@ public interface ProxyConfig
     // MikroCord start - allow Query response plugins populationg
     boolean isPopulateGS4QueryPlugins();
     // MikroCord end
+    // MikroCord start - add option to disable noisy packet decode exception
+    boolean isLogNoisyPacketDecodeException();
+    // MikroCord end
 }
diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
index a46bbc78..2bafd106 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/MinecraftDecoder.java
@@ -21,6 +21,9 @@ public class MinecraftDecoder extends MessageToMessageDecoder<ByteBuf>
     @Setter
     private boolean supportsForge = false;
     private final boolean allowEmptyPackets; // Waterfall
+    // MikroCord start - add option to disable noisy packet decode exception
+    private final boolean noisyPacketDecodeException;
+    // MikroCord end
 
     public MinecraftDecoder(Protocol protocol, boolean server, int protocolVersion) {
         // Waterfall start
@@ -32,6 +35,21 @@ public class MinecraftDecoder extends MessageToMessageDecoder<ByteBuf>
         this.server = server;
         this.protocolVersion = protocolVersion;
         this.allowEmptyPackets = allowEmptyPackets; // Waterfall
+        // MikroCord start - add option to disable noisy packet decode exception
+        boolean noisyPacketDecodeException;
+        try {
+            Class<?> configClass = Class.forName("eu.mikroskeem.mikrocord.conf.MikroCordConfiguration");
+            java.lang.reflect.Method instanceGetter = configClass.getMethod("getInstance");
+            Object configInstance = instanceGetter.invoke(null);
+
+            // Get config values
+            java.lang.reflect.Method isLogNoisyPacketDecodeException = configClass.getMethod("isLogNoisyPacketDecodeException");
+            noisyPacketDecodeException = (Boolean) isLogNoisyPacketDecodeException.invoke(configInstance);
+        } catch (Exception e) {
+            noisyPacketDecodeException = false;
+        }
+        this.noisyPacketDecodeException = noisyPacketDecodeException;
+        // MikroCord end
     }
 
     @Override
@@ -79,7 +97,16 @@ public class MinecraftDecoder extends MessageToMessageDecoder<ByteBuf>
             } else {
                 packetTypeStr = "unknown";
             }
+            /* // MikroCord start - add option to disable noisy packet decode exception
             throw new DecoderException("Error decoding packet " + packetTypeStr + " with contents:\n" + ByteBufUtil.prettyHexDump(slice), e);
+            */
+            DecoderException de = new DecoderException("Error decoding packet " + packetTypeStr + " with contents:\n" + ByteBufUtil.prettyHexDump(slice), e);
+            if (!this.noisyPacketDecodeException) {
+                de.setStackTrace(new StackTraceElement[0]);
+                de.getCause().setStackTrace(new StackTraceElement[0]);
+            }
+            throw de;
+            // MikroCord end
         } finally
         {
             if ( slice != null )
diff --git a/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
index e6a7d92b..30237777 100644
--- a/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
+++ b/proxy/src/main/java/eu/mikroskeem/mikrocord/conf/MikroCordConfiguration.java
@@ -20,6 +20,10 @@ import java.util.logging.Level;
  * @author Mark Vainomaa
  */
 public class MikroCordConfiguration extends WaterfallConfiguration {
+    public MikroCordConfiguration() {
+        instance = this;
+    }
+
     @Getter
     private String restartMessage = "Proxy is restarting";
 
@@ -53,6 +57,9 @@ public class MikroCordConfiguration extends WaterfallConfiguration {
     @Getter
     private boolean populateGS4QueryPlugins = false;
 
+    @Getter
+    private boolean logNoisyPacketDecodeException = false;
+
     @Override
     public void load() {
         super.load();
@@ -70,6 +77,7 @@ public class MikroCordConfiguration extends WaterfallConfiguration {
         allowTablistRewrite = config.getBoolean("performance.allow_tablist_rewrite", allowTablistRewrite);
         internScoreboardTeamStrings = config.getBoolean("performance.intern_scoreboard_team_strings", internScoreboardTeamStrings);
         populateGS4QueryPlugins = config.getBoolean("query.populate_plugins", populateGS4QueryPlugins);
+        logNoisyPacketDecodeException = config.getBoolean("log.noisy_packet_decode_exception", logNoisyPacketDecodeException);
     }
 
     private int setupTfo(int value) {
@@ -79,4 +87,7 @@ public class MikroCordConfiguration extends WaterfallConfiguration {
         }
         return actual.value;
     }
+
+    @Getter
+    private static MikroCordConfiguration instance;
 }
-- 
2.22.1

