From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Sun, 25 Aug 2019 04:54:53 +0300
Subject: [PATCH] Improve ProxyServer#matchPlayer


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
index a65b6d40c0e1e93aad499d2fcef2ee9ef7a7b7b6..710e346462182906f26e8b3db49d57d4eebc4fcc 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCord.java
@@ -843,6 +843,7 @@ public class BungeeCord extends ProxyServer
             return Collections.singleton( exactMatch );
         }
 
+        /* MikroCord start - improve matchPlayer
         return Sets.newHashSet( Iterables.filter( getPlayers(), new Predicate<ProxiedPlayer>()
         {
 
@@ -852,6 +853,19 @@ public class BungeeCord extends ProxyServer
                 return ( input == null ) ? false : input.getName().toLowerCase( Locale.ROOT ).startsWith( partialName.toLowerCase( Locale.ROOT ) );
             }
         } ) );
+        */
+        String lowerPartialMatch = partialName.toLowerCase(Locale.ROOT);
+        var foundPlayers = new java.util.LinkedList<ProxiedPlayer>();
+        for (var player : getPlayers()) {
+            if (player == null) { // Should not happen though?
+                continue;
+            }
+            if (player.getName().toLowerCase(Locale.ROOT).startsWith(lowerPartialMatch)) {
+                foundPlayers.add(player);
+            }
+        }
+        return java.util.Set.copyOf(foundPlayers);
+        // MikroCord end
     }
 
     @Override
