From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Thu, 6 Feb 2020 17:53:53 +0200
Subject: [PATCH] Print out Java runtime and OS information on startup


diff --git a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
index adbe8f37ecd8421843bc490916a14e6b848c4c4f..d5b050cc2c2fc0572a4b4adb254e09939da348a9 100644
--- a/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
+++ b/proxy/src/main/java/net/md_5/bungee/BungeeCordLauncher.java
@@ -61,6 +61,23 @@ public class BungeeCordLauncher
         }
 
         BungeeCord bungee = new BungeeCord();
+        // Firefly start - Output JVM and OS information on startup
+        // Credit goes to Paper
+        var runtimeMX = java.lang.management.ManagementFactory.getRuntimeMXBean();
+        var osMX = java.lang.management.ManagementFactory.getOperatingSystemMXBean();
+        if (runtimeMX != null && osMX != null) {
+            bungee.getLogger().log(
+                    java.util.logging.Level.INFO,
+                    "System Info: Java: {0} ({1} {2}) Host: {3} {4} ({5})",
+                    new Object[]{
+                            runtimeMX.getSpecVersion(), runtimeMX.getVmName(), runtimeMX.getVmVersion(),
+                            osMX.getName(), osMX.getVersion(), osMX.getArch()
+                    }
+            );
+        } else {
+            bungee.getLogger().warning("Unable to read system info");
+        }
+        // Firefly end
         ProxyServer.setInstance( bungee );
         bungee.getLogger().info( "Enabled Firefly version " + bungee.getVersion() );
         bungee.start();
