From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Thu, 26 Sep 2019 16:58:24 +0300
Subject: [PATCH] Harden against malicious clients


diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index ca7d5fd53fb242d3d8dc008f6f1b7b79da1f2fd1..c70bf92054b3388b0f8f31d515f56335c96c1011 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -334,6 +334,10 @@ public class InitialHandler extends PacketHandler implements PendingConnection
                 }
                 break;
             default:
+                // MikroCord start - harden
+                bungee.getLogger().log(Level.WARNING, "{0} Invalid protocol {1}, disconnecting", new Object[] { this, handshake.getRequestedProtocol() });
+                ch.close();
+                if (false) // MikroCord end
                 throw new IllegalArgumentException( "Cannot request protocol " + handshake.getRequestedProtocol() );
         }
     }
@@ -344,7 +348,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
         Preconditions.checkState( thisState == State.USERNAME, "Not expecting USERNAME" );
         this.loginRequest = loginRequest;
 
-        if ( getName().contains( "." ) )
+        if ( getName().contains( "." ) || getName().contains(" ") ) // MikroCord - harden
         {
             disconnect( bungee.getTranslation( "name_invalid" ) );
             return;
