From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Mon, 27 Jan 2020 21:56:06 +0200
Subject: [PATCH] Allow disabling legacy ping handling


diff --git a/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java b/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
index 0757f3bdfed36e5c2b71c01f847ecff56e409fbe..55d42b9775ec274b21568a86303f0a9e0bbde1f5 100644
--- a/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
+++ b/api/src/main/java/eu/mikroskeem/firefly/api/config/FireflyProxyConfig.java
@@ -94,4 +94,6 @@ public interface FireflyProxyConfig {
     boolean isDropConnectionsSendingInvalidPackets();
 
     boolean isLogPacketDecodeErrors();
+
+    boolean isHandleLegacyPing();
 }
diff --git a/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java b/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
index 85d48589a7d604c022ded8299e154eaa31e0130b..30788138b01d3da1cb3bd76d66b6a85f2f0253f6 100644
--- a/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
+++ b/proxy/src/main/java/eu/mikroskeem/firefly/conf/FireflyConfiguration.java
@@ -88,6 +88,9 @@ public class FireflyConfiguration extends WaterfallConfiguration {
     @Getter
     private boolean logUnparsableChatJson = true;
 
+    @Getter
+    private boolean handleLegacyPing = true;
+
     @Override
     public void load() {
         super.load();
@@ -114,6 +117,7 @@ public class FireflyConfiguration extends WaterfallConfiguration {
         logPacketDecodeErrors = config.getBoolean("packets.log_packet_decode_errors", logPacketDecodeErrors);
         logUnparsableChatJson = config.getBoolean("log.unparsable_chat_json", logUnparsableChatJson);
         ComponentSerializer.debugJsonParsing.set(logUnparsableChatJson);
+        handleLegacyPing = config.getBoolean("packets.handle_legacy_ping", handleLegacyPing);
     }
 
     private TCPFastOpenMode setupTfo(int value) {
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index 2a8b1e5421f4a0fa29145be812d436ad921bd87e..50893057dacd94d58f331869594c45cd3914b3aa 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -182,6 +182,12 @@ public class InitialHandler extends PacketHandler implements PendingConnection
     {
         this.legacy = true;
         final boolean v1_5 = ping.isV1_5();
+        // Firefly start
+        if (!bungee.config.isHandleLegacyPing()) {
+            ch.close();
+            return;
+        }
+        // Firefly end
 
         ServerPing legacy = new ServerPing( new ServerPing.Protocol( bungee.getName() + " " + bungee.getGameVersion(), bungee.getProtocolVersion() ),
                 new ServerPing.Players( listener.getMaxPlayers(), bungee.getOnlineCount(), null ),
