From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Fri, 6 Mar 2020 00:24:43 +0200
Subject: [PATCH] Fire ClientConnectEvent later when PROXY protocol is enabled


diff --git a/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java b/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
index 2cf0d588eb78fb724bf7a15f4656dfd1d92ad10d..040c9b3a7cc0576f6464aee1cc64e087cfc0113d 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/HandlerBoss.java
@@ -82,6 +82,12 @@ public class HandlerBoss extends ChannelInboundHandlerAdapter
         {
             HAProxyMessage proxy = (HAProxyMessage) msg;
             InetSocketAddress newAddress = new InetSocketAddress( proxy.sourceAddress(), proxy.sourcePort() );
+            // Firefly start - Fire ClientConnectEvent later when PROXY protocol is enabled
+            if (!new net.md_5.bungee.api.event.ClientConnectEvent(newAddress, channel.getHandle().attr(PipelineUtils.LISTENER).get()).callEvent()) {
+                ctx.channel().close();
+                return;
+            }
+            // Firefly end
 
             ProxyServer.getInstance().getLogger().log( Level.FINE, "Set remote address via PROXY {0} -> {1}", new Object[]
             {
diff --git a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
index 9b3c9ed5192a02b3e3dc91e58d878ea3a438f2c8..e3ea4a1f7c868183d5ed0fb05a94f41e5ffd6499 100644
--- a/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
+++ b/proxy/src/main/java/net/md_5/bungee/netty/PipelineUtils.java
@@ -67,7 +67,7 @@ public class PipelineUtils
             }
             ListenerInfo listener = ch.attr( LISTENER ).get();
 
-            if ( BungeeCord.getInstance().getPluginManager().callEvent( new ClientConnectEvent( remoteAddress, listener ) ).isCancelled() )
+            if (!listener.isProxyProtocol() && BungeeCord.getInstance().getPluginManager().callEvent( new ClientConnectEvent( remoteAddress, listener ) ).isCancelled() ) // Firefly - Fire ClientConnectEvent later when PROXY protocol is enabled
             {
                 ch.close();
                 return;
