From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Mon, 16 Mar 2020 12:21:48 +0200
Subject: [PATCH] Avoid expensive hostname lookups

Pointed out by LinsaFTW in SpigotMC/BungeeCord#2719

diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 20dfe6153dd6a5c6102049a5cd3dd55738d5ccb8..4411e187fd4c145188bd3aee5bc7d3b348852fe3 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -529,7 +529,7 @@ public class DownstreamBridge extends PacketHandler
                 {
                     out.writeUTF( "ServerIP" );
                     out.writeUTF( info.getName() );
-                    out.writeUTF( info.getAddress().getAddress().getHostAddress() );
+                    out.writeUTF( info.getAddress().getHostString() ); // Firefly - avoid expensive hostname lookups
                     out.writeShort( info.getAddress().getPort() );
                 }
             }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index 52cfebea688844e44ec24602249b6fb7fd52568c..a6bd92351f8a5c9ef9ff7753789078404a72907c 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -520,7 +520,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
         }
         String encodedHash = URLEncoder.encode( new BigInteger( sha.digest() ).toString( 16 ), "UTF-8" );
 
-        String preventProxy = ( BungeeCord.getInstance().config.isPreventProxyConnections() && getSocketAddress() instanceof InetSocketAddress ) ? "&ip=" + URLEncoder.encode( getAddress().getAddress().getHostAddress(), "UTF-8" ) : "";
+        String preventProxy = ( BungeeCord.getInstance().config.isPreventProxyConnections() && getSocketAddress() instanceof InetSocketAddress ) ? "&ip=" + URLEncoder.encode( getAddress().getHostString(), "UTF-8" ) : ""; // Firefly - avoid expensive hostname lookups
         String authURL = String.format( MOJANG_AUTH_URL, encName, encodedHash, preventProxy );
 
         Callback<String> handler = new Callback<String>()
