From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Mon, 16 Mar 2020 12:21:48 +0200
Subject: [PATCH] Avoid expensive hostname lookups

Pointed out by LinsaFTW in SpigotMC/BungeeCord#2719

diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index c27000045c77f33cc01b57b11e9cefd8a69436a8..b52d9b26dbee439da13029a0369c0d54b6700571 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -534,7 +534,7 @@ public class DownstreamBridge extends PacketHandler
                 {
                     out.writeUTF( "ServerIP" );
                     out.writeUTF( info.getName() );
-                    out.writeUTF( info.getAddress().getAddress().getHostAddress() );
+                    out.writeUTF( info.getAddress().getHostString() ); // Firefly - avoid expensive hostname lookups
                     out.writeShort( info.getAddress().getPort() );
                 }
             }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index 108b0fee4212e1e684a1e531b90760f4f46cd4be..d0d30b33795541298b18217884cb0325ab1de3aa 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -520,7 +520,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
         }
         String encodedHash = URLEncoder.encode( new BigInteger( sha.digest() ).toString( 16 ), "UTF-8" );
 
-        String preventProxy = ( BungeeCord.getInstance().config.isPreventProxyConnections() && getSocketAddress() instanceof InetSocketAddress ) ? "&ip=" + URLEncoder.encode( getAddress().getAddress().getHostAddress(), "UTF-8" ) : "";
+        String preventProxy = ( BungeeCord.getInstance().config.isPreventProxyConnections() && getSocketAddress() instanceof InetSocketAddress ) ? "&ip=" + URLEncoder.encode( getAddress().getHostString(), "UTF-8" ) : ""; // Firefly - avoid expensive hostname lookups
         String authURL = String.format( MOJANG_AUTH_URL, encName, encodedHash, preventProxy );
 
         Callback<String> handler = new Callback<String>()
