From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Mon, 16 Mar 2020 12:21:48 +0200
Subject: [PATCH] Avoid expensive hostname lookups

Pointed out by LinsaFTW in SpigotMC/BungeeCord#2719

diff --git a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
index 06a7e14286d296c2fd32c507b31985c58f43a80a..b37d5f57e093075a9e4c4cb0d215d7797314742a 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/DownstreamBridge.java
@@ -505,7 +505,7 @@ public class DownstreamBridge extends PacketHandler
                 {
                     out.writeUTF( "ServerIP" );
                     out.writeUTF( info.getName() );
-                    out.writeUTF( info.getAddress().getAddress().getHostAddress() );
+                    out.writeUTF( info.getAddress().getHostString() ); // Firefly - avoid expensive hostname lookups
                     out.writeShort( info.getAddress().getPort() );
                 }
             }
diff --git a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
index 136fcfc5fbdd16ca7b04c79a74260adfa0ad4beb..e40f438ba840afcd75454e9477c1b71febb5835f 100644
--- a/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
+++ b/proxy/src/main/java/net/md_5/bungee/connection/InitialHandler.java
@@ -518,7 +518,7 @@ public class InitialHandler extends PacketHandler implements PendingConnection
         }
         String encodedHash = URLEncoder.encode( new BigInteger( sha.digest() ).toString( 16 ), "UTF-8" );
 
-        String preventProxy = ( BungeeCord.getInstance().config.isPreventProxyConnections() && getSocketAddress() instanceof InetSocketAddress ) ? "&ip=" + URLEncoder.encode( getAddress().getAddress().getHostAddress(), "UTF-8" ) : "";
+        String preventProxy = ( BungeeCord.getInstance().config.isPreventProxyConnections() && getSocketAddress() instanceof InetSocketAddress ) ? "&ip=" + URLEncoder.encode( getAddress().getHostString(), "UTF-8" ) : ""; // Firefly - avoid expensive hostname lookups
         String authURL = "https://sessionserver.mojang.com/session/minecraft/hasJoined?username=" + encName + "&serverId=" + encodedHash + preventProxy;
 
         Callback<String> handler = new Callback<String>()
