From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Mark Vainomaa <mikroskeem@mikroskeem.eu>
Date: Wed, 24 Jun 2020 16:32:17 +0300
Subject: [PATCH] 1.16.1 support


diff --git a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
index c959282109955d74380ef784e480843ba0d4f590..f0c2795cdd8afd540cf4286138018ab2dafeaa47 100644
--- a/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
+++ b/protocol/src/main/java/net/md_5/bungee/protocol/ProtocolConstants.java
@@ -29,6 +29,7 @@ public class ProtocolConstants
     public static final int MINECRAFT_1_15_1 = 575;
     public static final int MINECRAFT_1_15_2 = 578;
     public static final int MINECRAFT_1_16 = 735;
+    public static final int MINECRAFT_1_16_1 = 736; // Firefly
     public static final List<String> SUPPORTED_VERSIONS = Arrays.asList(
             "1.8.x",
             "1.9.x",
@@ -63,7 +64,8 @@ public class ProtocolConstants
             ProtocolConstants.MINECRAFT_1_15,
             ProtocolConstants.MINECRAFT_1_15_1,
             ProtocolConstants.MINECRAFT_1_15_2,
-            ProtocolConstants.MINECRAFT_1_16
+            ProtocolConstants.MINECRAFT_1_16,
+            ProtocolConstants.MINECRAFT_1_16_1 // Firefly
     );
     // Firefly start
     public static final int LOWEST_SUPPORTED_VERSION_ID;
diff --git a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
index 5b76ee070e856f59f81af9667d5b880b08fa8b65..ca271c4d371b4d655f12c6cd76f042293c770a9f 100644
--- a/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
+++ b/proxy/src/main/java/net/md_5/bungee/entitymap/EntityMap.java
@@ -68,6 +68,7 @@ public abstract class EntityMap
             case ProtocolConstants.MINECRAFT_1_15_2:
                 return EntityMap_1_15.INSTANCE;
             case ProtocolConstants.MINECRAFT_1_16:
+            case ProtocolConstants.MINECRAFT_1_16_1: // Firefly
                 return EntityMap_1_16.INSTANCE;
         }
         throw new RuntimeException( "Version " + version + " has no entity map" );
